

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>CIF Parser &mdash; Gemmi 0.3.7 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Symmetry" href="symmetry.html" />
    <link rel="prev" title="Installation" href="install.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Gemmi
          

          
          </a>

          
            
            
              <div class="version">
                0.3.7
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">CIF Parser</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#what-are-star-cif-ddl-mmcif">What are STAR, CIF, DDL, mmCIF?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#what-is-parsed">What is parsed?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#getting-started">Getting started</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#c">C++</a></li>
<li class="toctree-l3"><a class="reference internal" href="#python">Python</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#dom-and-sax">DOM and SAX</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1">C++</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2">Python</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#reading-a-file">Reading a file</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id3">C++</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id4">Python</a></li>
<li class="toctree-l3"><a class="reference internal" href="#low-level-functions">Low-level functions</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#writing-a-file">Writing a file</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id5">C++</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id6">Python</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#document">Document</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id7">C++</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id8">Python</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#block">Block</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id9">C++</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id10">Python</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#frame">Frame</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pairs-and-loops">Pairs and Loops</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id11">C++</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id12">Python</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#column">Column</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id13">C++</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id14">Python</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#table">Table</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#row-wise-access">Row-wise access</a></li>
<li class="toctree-l3"><a class="reference internal" href="#column-wise-access">Column-wise access</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mmcif-categories">mmCIF categories</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id15">C++</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id16">Python</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#json">JSON</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id17">C++</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id18">Python</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#design-choices">Design choices</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#parser">Parser</a></li>
<li class="toctree-l3"><a class="reference internal" href="#data-structures">Data structures</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#performance">Performance</a></li>
<li class="toctree-l2"><a class="reference internal" href="#directory-walking">Directory walking</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id19">C++</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id20">Python</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#examples">Examples</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#mmcif-to-xyz">mmCIF to XYZ</a></li>
<li class="toctree-l3"><a class="reference internal" href="#auth-vs-label">auth vs label</a></li>
<li class="toctree-l3"><a class="reference internal" href="#amino-acid-frequency">Amino acid frequency</a></li>
<li class="toctree-l3"><a class="reference internal" href="#custom-pdb-search">Custom PDB search</a></li>
<li class="toctree-l3"><a class="reference internal" href="#search-pdb-by-elements">Search PDB by elements</a></li>
<li class="toctree-l3"><a class="reference internal" href="#solvent-content-vs-resolution">Solvent content vs resolution</a></li>
<li class="toctree-l3"><a class="reference internal" href="#weights">Weights</a></li>
<li class="toctree-l3"><a class="reference internal" href="#disulfide-bonds">Disulfide bonds</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mmjson-like-data">mmJSON-like data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#chemical-component-dictionary">Chemical Component Dictionary</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="symmetry.html">Symmetry</a></li>
<li class="toctree-l1"><a class="reference internal" href="mol.html">Molecular models</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis.html">Structure analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="grid.html">Grids and maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="hkl.html">Reciprocal space</a></li>
<li class="toctree-l1"><a class="reference internal" href="utils.html">Gemmi program</a></li>
<li class="toctree-l1"><a class="reference external" href="https://project-gemmi.github.io/gemmi/api/python">Python API</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Gemmi</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>CIF Parser</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/cif.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="cif-parser">
<h1>CIF Parser<a class="headerlink" href="#cif-parser" title="Permalink to this headline">¶</a></h1>
<p>This section covers working with CIF files on a syntactic level.</p>
<p>Higher-level functions that understand semantics of:</p>
<ul class="simple">
<li><p>small molecule or inorganic CIF files,</p></li>
<li><p>macromolecular PDBx/mmCIF,</p></li>
<li><p>and monomer/ligand cif files as used for macromolecular restraints</p></li>
</ul>
<p>are documented in section <a class="reference internal" href="mol.html#molecular"><span class="std std-ref">Molecular models</span></a>.</p>
<div class="section" id="what-are-star-cif-ddl-mmcif">
<span id="cif-intro"></span><h2>What are STAR, CIF, DDL, mmCIF?<a class="headerlink" href="#what-are-star-cif-ddl-mmcif" title="Permalink to this headline">¶</a></h2>
<p>(in case someone comes here when looking for a serialization format)</p>
<p>STAR is a human-readable data serialization format (think XML or JSON)
that happens to be known and used only in molecular-structure sciences.</p>
<p>CIF (Crystallographic Information File) – a file format used
in crystallography – is a restricted derivative of STAR.
It is restricted in features (to make implementation easier),
but also imposes arbitrary limits – for example on the line length.</p>
<p>DDL is a schema language for STAR/CIF.</p>
<p>All of them (STAR, CIF and DDL) have multiple versions.
We will be more specific in the following sections.</p>
<p>The STAR/CIF syntax is relatively simple, but may be confusing at first.
(Note that the initial version of STAR was published by Sydney Hall in 1991 –
before XML and long before JSON and YAML, not to mention TOML).</p>
<p>Key-value pairs have the form:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">_year</span> <span class="mi">2017</span>  <span class="c1"># means year = 2017</span>
</pre></div>
</div>
<p>Blocks (sections) begin with the <code class="docutils literal notranslate"><span class="pre">data_</span></code> keyword
with a block name glued to it:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">data_tomato</span>   <span class="c1"># [tomato]</span>
<span class="n">_color</span> <span class="n">red</span>    <span class="c1"># color = &quot;red&quot;</span>
</pre></div>
</div>
<p>Importantly, unlike XML/JSON/YAML/TOML, STAR is designed
to concisely represent tabular data. This example from TOML docs:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># TOML</span>
<span class="n">points</span> <span class="o">=</span> <span class="p">[</span> <span class="p">{</span> <span class="n">x</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">z</span> <span class="o">=</span> <span class="mi">3</span> <span class="p">},</span>
           <span class="p">{</span> <span class="n">x</span> <span class="o">=</span> <span class="mi">7</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span> <span class="n">z</span> <span class="o">=</span> <span class="mi">9</span> <span class="p">},</span>
           <span class="p">{</span> <span class="n">x</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="n">z</span> <span class="o">=</span> <span class="mi">8</span> <span class="p">}</span> <span class="p">]</span>
</pre></div>
</div>
<p>could be expressed as a so-called <em>loop</em> in STAR (keyword <code class="docutils literal notranslate"><span class="pre">loop_</span></code>
followed by column names followed by values):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># STAR/CIF</span>
<span class="n">loop_</span>
<span class="n">_points</span><span class="o">.</span><span class="n">x</span> <span class="n">_points</span><span class="o">.</span><span class="n">y</span> <span class="n">_points</span><span class="o">.</span><span class="n">z</span>
<span class="mi">1</span> <span class="mi">2</span> <span class="mi">3</span>
<span class="mi">7</span> <span class="mi">8</span> <span class="mi">9</span>
<span class="mi">2</span> <span class="mi">4</span> <span class="mi">8</span>
</pre></div>
</div>
<p>Typically, long tables (loops) make most of the CIF content:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>1    N N   . LEU A 11  ? 0.5281 0.5618 0.5305 -0.0327 -0.0621 0.0104
2    C CA  . LEU A 11  ? 0.5446 0.5722 0.5396 -0.0317 -0.0632 0.0080
# many, many lines...
5331 S SD  . MET C 238 ? 2.2952 2.3511 2.3275 -0.0895 0.0372  -0.0230
5332 C CE  . MET C 238 ? 1.5699 1.6247 1.6108 -0.0907 0.0388  -0.0244
</pre></div>
</div>
<p>The dot and question mark in the example above are two null types.
In the CIF spec: <code class="docutils literal notranslate"><span class="pre">?</span></code> = <em>unknown</em> and <code class="docutils literal notranslate"><span class="pre">.</span></code> = <em>not applicable</em>.
In mmCIF files <code class="docutils literal notranslate"><span class="pre">.</span></code> is used for mandatory items, <code class="docutils literal notranslate"><span class="pre">?</span></code> for not mandatory.</p>
<p>The CIF syntax has a serious flaw resulting from historical trade-offs:
a string that can be interpreted as a number does not need to be quoted.
Therefore, the type of <code class="docutils literal notranslate"><span class="pre">5332</span></code> above is not certain:
the JSON equivalent can be either <code class="docutils literal notranslate"><span class="pre">5332</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;5332&quot;</span></code>.</p>
<p>Note: “STAR File” is trademarked by IUCr, and it used to be <a class="reference external" href="https://patents.google.com/patent/WO1991016682A1">patented</a>.</p>
<p>The mmCIF format (by mmCIF we mean what is more formally called PDBx/mmCIF)
is the CIF syntax + a huge dictionary (ontology/schema) in DDL2.
The dictionary defines relations between columns in different tables,
which makes it resemble a relational database (it was designed at the
height of popularity of RDBMSs).</p>
<p><strong>Where are the specs?</strong></p>
<p>International Tables for Crystallography
<a class="reference external" href="http://it.iucr.org/Ga/contents/">Vol. G (2006)</a>
describes all of the STAR, CIF 1.1, DDL1 and DDL2.
If you don’t have access to it – IUCr website has specs of
<a class="reference external" href="http://www.iucr.org/resources/cif/spec/version1.1">CIF1.1</a>
and <a class="reference external" href="http://www.iucr.org/resources/cif/ddl">DDLs</a>.
As far as I can tell all versions of the STAR spec are behind paywalls.</p>
<p>Later versions of the formats (hardly used as of 2017)
are described in these articles:
<a class="reference external" href="https://dx.doi.org/10.1021/ci300074v">STAR</a> (2012),
<a class="reference external" href="http://pubs.acs.org/doi/abs/10.1021/ci300075z">DDLm</a> and
<a class="reference external" href="http://pubs.acs.org/doi/abs/10.1021/ci300076w">dREL</a> (2012),
and <a class="reference external" href="http://journals.iucr.org/j/issues/2016/01/00/aj5269/">CIF 2.0</a> (2016).
Only the last one is freely available.</p>
<p>PDBx/mmCIF is documented at <a class="reference external" href="http://mmcif.pdb.org/">mmcif.pdb.org</a>.</p>
</div>
<div class="section" id="what-is-parsed">
<h2>What is parsed?<a class="headerlink" href="#what-is-parsed" title="Permalink to this headline">¶</a></h2>
<p>The parser supports CIF 1.1 spec and some extras.</p>
<p>Currently, it is available as:</p>
<ul class="simple">
<li><p>C++11 header-only library, and</p></li>
<li><p>Python (2 and 3) extension module.</p></li>
</ul>
<p>We use it to read:</p>
<ul class="simple">
<li><p>mmCIF files (both coordinates and structure factors)</p></li>
<li><p>CIF files from Crystallography Open Database (COD)</p></li>
<li><p>Chemical Component Dictionary from PDB</p></li>
<li><p>DDL1 and DDL2 dictionaries from IUCr and PDB</p></li>
<li><p>monomer library a.k.a. Refmac dictionary</p></li>
</ul>
<p>The parser handles:</p>
<ul class="simple">
<li><p>all constructs of CIF 1.1 (including <em>save frames</em>),</p></li>
<li><p>the <code class="docutils literal notranslate"><span class="pre">global_</span></code> and <code class="docutils literal notranslate"><span class="pre">stop_</span></code> keywords from STAR – needed for Refmac
monomer library and <code class="docutils literal notranslate"><span class="pre">mmcif_nmr-star.dic</span></code>, respectively.</p></li>
</ul>
<p>It could be extended to handle also the new features of CIF 2.0
or even the full STAR format, but we don’t have a good reason to do this.
The same goes for DDLm/dREL.</p>
<p>The parser does not handle CIF 1.1 conventions, as they are not part
of the syntax:
line wrapping (<code class="docutils literal notranslate"><span class="pre">eol;\eol</span></code>),
Greek letters (<code class="docutils literal notranslate"><span class="pre">\m</span></code> -&gt; µ),
accented letters (<code class="docutils literal notranslate"><span class="pre">\'o</span></code> -&gt; ó),
special alphabetic characters (<code class="docutils literal notranslate"><span class="pre">\%A</span></code> -&gt; Å)
and other codes (<code class="docutils literal notranslate"><span class="pre">\\infty</span></code> -&gt; ∞).</p>
<p>CIF parsers in the small-molecules field need to deal with incorrect syntax.
The papers about <a class="reference external" href="https://doi.org/10.1107/S0021889811041161">iotbx.cif</a>
and <a class="reference external" href="http://dx.doi.org/10.1107/S1600576715022396">COD::CIF::Parser</a>
enumerate 12 and 10 common error categories, respectively.
Nowadays the problem is less severe, especially in the MX community
that embraced the CIF format later. So we’ve decided to add only
the following rules to relax the syntax:</p>
<ul class="simple">
<li><p>names and lines can have any length like in STAR
(the CIF spec imposes the limit of 2048 characters, but some mmCIF files
from PDB exceed it, e.g. 3j3q.cif),</p></li>
<li><p>quoted strings may contain non-ascii characters (if nothing has changed
one entry in the PDB has byte A0 corresponding to non-breaking space),</p></li>
<li><p>a table (loop) can have no values if it is followed by another loop
or block end (because Refmac monomer library),</p></li>
<li><p>block name (<em>blockcode</em>) can be empty, i.e. the block can start
with bare <code class="docutils literal notranslate"><span class="pre">data_</span></code> keyword (RELION writes such files),</p></li>
<li><p>unquoted strings cannot start with keywords (STAR spec is ambiguous
about this – see
<a class="reference external" href="http://www.globalphasing.com/startools/">StarTools doc</a> for details;
this rule is actually about simplifying not relaxing),</p></li>
<li><p>missing value in a key-value pair is optionally allowed
if whitespace after the tag ends with a new-line character.
More specifically, the parsing step allows for missing value in such case,
but the next validation step (which can be skipped when using using
low-level functions, such as <code class="docutils literal notranslate"><span class="pre">parse_file()</span></code>) throws an error.</p></li>
</ul>
</div>
<div class="section" id="getting-started">
<h2>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this headline">¶</a></h2>
<div class="section" id="c">
<h3>C++<a class="headerlink" href="#c" title="Permalink to this headline">¶</a></h3>
<p>CIF parser is implemented in header files,
so you do not need to compile Gemmi.
It has a single dependency: PEGTL (also header-only),
which is included under the <code class="docutils literal notranslate"><span class="pre">include/gemmi/third_party</span></code> directory.
All you need is to make sure that Gemmi headers are in your
project’s include path, and compile your program as C++11 or later.</p>
<p>Let us start with a simple example.
This little program reads mmCIF file and shows weights of the chemical
components:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="cp">#include</span> <span class="cpf">&lt;iostream&gt;</span><span class="cp"></span>
<span class="cp">#include</span> <span class="cpf">&lt;gemmi/cif.hpp&gt;</span><span class="cp"></span>

<span class="k">namespace</span> <span class="n">cif</span> <span class="o">=</span> <span class="n">gemmi</span><span class="o">::</span><span class="n">cif</span><span class="p">;</span>

<span class="kt">int</span> <span class="nf">main</span><span class="p">()</span> <span class="p">{</span>
  <span class="n">cif</span><span class="o">::</span><span class="n">Document</span> <span class="n">doc</span> <span class="o">=</span> <span class="n">cif</span><span class="o">::</span><span class="n">read_file</span><span class="p">(</span><span class="s">&quot;1mru.cif&quot;</span><span class="p">);</span>
  <span class="k">for</span> <span class="p">(</span><span class="n">cif</span><span class="o">::</span><span class="n">Block</span><span class="o">&amp;</span> <span class="nl">block</span> <span class="p">:</span> <span class="n">doc</span><span class="p">.</span><span class="n">blocks</span><span class="p">)</span>
    <span class="k">for</span> <span class="p">(</span><span class="k">auto</span> <span class="nl">cc</span> <span class="p">:</span> <span class="n">block</span><span class="p">.</span><span class="n">find</span><span class="p">(</span><span class="s">&quot;_chem_comp.&quot;</span><span class="p">,</span> <span class="p">{</span><span class="s">&quot;id&quot;</span><span class="p">,</span> <span class="s">&quot;formula_weight&quot;</span><span class="p">}))</span>
      <span class="n">std</span><span class="o">::</span><span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="n">cc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;&lt;</span> <span class="s">&quot; weights &quot;</span> <span class="o">&lt;&lt;</span> <span class="n">cc</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;&lt;</span> <span class="n">std</span><span class="o">::</span><span class="n">endl</span><span class="p">;</span>
<span class="p">}</span>
</pre></div>
</div>
<p>To compile it on Unix system you need to fetch Gemmi source code
and run a compiler:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>git clone https://github.com/project-gemmi/gemmi.git
c++ -std=c++11 -Igemmi/include -O2 my_program.cpp
</pre></div>
</div>
</div>
<div class="section" id="python">
<h3>Python<a class="headerlink" href="#python" title="Permalink to this headline">¶</a></h3>
<p>Python module for Python 2.7 and 3.x
can be installed with pip, as described in the
<a class="reference internal" href="install.html#install-py"><span class="std std-ref">Installation</span></a> section.
After installation <code class="docutils literal notranslate"><span class="pre">pydoc</span> <span class="pre">gemmi.cif</span></code> should list all classes and methods.</p>
<p>To start with a simple example, here is a program that says hello to each
element found in mmCIF:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="hll"><span class="kn">from</span> <span class="nn">gemmi</span> <span class="kn">import</span> <span class="n">cif</span>
</span>
<span class="n">greeted</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
    <span class="k">try</span><span class="p">:</span>
<span class="hll">        <span class="n">doc</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">read_file</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>  <span class="c1"># copy all the data from mmCIF file</span>
</span><span class="hll">        <span class="n">block</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">sole_block</span><span class="p">()</span>  <span class="c1"># mmCIF has exactly one block</span>
</span><span class="hll">        <span class="k">for</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">block</span><span class="o">.</span><span class="n">find_loop</span><span class="p">(</span><span class="s2">&quot;_atom_site.type_symbol&quot;</span><span class="p">):</span>
</span>            <span class="k">if</span> <span class="n">element</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">greeted</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Hello &quot;</span> <span class="o">+</span> <span class="n">element</span><span class="p">)</span>
                <span class="n">greeted</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">element</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Oops. </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">e</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<p>More complex examples are shown in the <a class="reference internal" href="#cif-examples"><span class="std std-ref">Examples</span></a> section.</p>
<p>Internally, Python bindings use
<a class="reference external" href="https://github.com/pybind/pybind11">pybind11</a>.</p>
</div>
</div>
<div class="section" id="dom-and-sax">
<h2>DOM and SAX<a class="headerlink" href="#dom-and-sax" title="Permalink to this headline">¶</a></h2>
<p>The terms DOM and SAX originate from XML parsing, but they became also
names of general parsing styles.
Gemmi can parse CIF files in two ways that correspond to DOM and SAX:</p>
<ul class="simple">
<li><p>The usual way is to parse a file or a string into a document (DOM)
that can be easily accessed and manipulated.</p></li>
<li><p>Alternatively, from C++ only, one can define own
<a class="reference external" href="https://github.com/taocpp/PEGTL/blob/master/doc/Actions-and-States.md">PEGTL Actions</a>
for to the grammar rules from <code class="docutils literal notranslate"><span class="pre">cif.hpp</span></code>.
These actions will be triggered while reading a CIF file.</p></li>
</ul>
<p>This documention covers the DOM parsing only.
The hierarchy in the DOM reflects the structure of CIF 1.1:</p>
<ul class="simple">
<li><p>Document contains blocks.</p></li>
<li><p>Block can contain name-value pairs, loops and frames.</p></li>
<li><p>Frame can contain name-value pairs and loops.</p></li>
<li><p>Loop (<em>m</em>×<em>n</em> table) contains <em>n</em> column names and <em>m</em>×<em>n</em> values.</p></li>
</ul>
<p>Names are often called <em>tags</em>. The leading <code class="docutils literal notranslate"><span class="pre">_</span></code> is usually treated
as part of the tag, not just a syntactic feature. So we store tag string
with the underscore (<code class="docutils literal notranslate"><span class="pre">_my_tag</span></code>), and function that take tags as arguments
expect strings starting with <code class="docutils literal notranslate"><span class="pre">_</span></code>.
The case of tags is preserved.</p>
<p>Values have types. CIF 1.1 defines four base types:</p>
<ul class="simple">
<li><p>char (string)</p></li>
<li><p>uchar (ughh.. case-insensitive string)</p></li>
<li><p>numb (number that cannot be recognized as number on the syntax level)</p></li>
<li><p>null (one of two possible nulls: <code class="docutils literal notranslate"><span class="pre">?</span></code> and <code class="docutils literal notranslate"><span class="pre">.</span></code>)</p></li>
</ul>
<p>Additionally, DDL2 dictionaries specify subtypes.
For example, <code class="docutils literal notranslate"><span class="pre">int</span></code> and <code class="docutils literal notranslate"><span class="pre">float</span></code> are mmCIF subtypes of <code class="docutils literal notranslate"><span class="pre">numb</span></code>.</p>
<p>Since in general it is not possible to infer type without
the corresponding dictionary, the DOM stores raw strings (including quotes).
They can be later converted to required type using the following helper
functions:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">as_string()</span></code> – gets unquoted string,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">as_number()</span></code> – gets floating-point number,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">as_int()</span></code> – gets integer,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">as_char()</span></code> – gets single character,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">is_null()</span></code> – check if the value is null (i.e. <code class="docutils literal notranslate"><span class="pre">?</span></code> or <code class="docutils literal notranslate"><span class="pre">.</span></code>),</p></li>
</ul>
<p>and we have also</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">quote()</span></code> – the opposite of <code class="docutils literal notranslate"><span class="pre">as_string()</span></code> – add quotes appropriate
for the content of the string (usually, no quotes are necessary and no
quotes are added).</p></li>
</ul>
<div class="section" id="id1">
<h3>C++<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>All these helper functions are defined in <code class="docutils literal notranslate"><span class="pre">gemmi/cifdoc.hpp</span></code>
except for <code class="docutils literal notranslate"><span class="pre">as_number()</span></code> which is in <code class="docutils literal notranslate"><span class="pre">gemmi/numb.hpp</span></code>.
They take a string as an argument and work as expected, for example:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="kt">double</span> <span class="n">rfree</span> <span class="o">=</span> <span class="n">cif</span><span class="o">::</span><span class="n">as_number</span><span class="p">(</span><span class="n">raw_rfree_string</span><span class="p">);</span> <span class="c1">// NaN if it&#39;s &#39;?&#39; or &#39;.&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="id2">
<h3>Python<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">gemmi</span> <span class="kn">import</span> <span class="n">cif</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cif</span><span class="o">.</span><span class="n">as_number</span><span class="p">(</span><span class="s1">&#39;123&#39;</span><span class="p">)</span>
<span class="go">123.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cif</span><span class="o">.</span><span class="n">as_int</span><span class="p">(</span><span class="s1">&#39;123&#39;</span><span class="p">)</span>
<span class="go">123</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cif</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="s1">&#39;two words&#39;</span><span class="p">)</span>
<span class="go">&quot;&#39;two words&#39;&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cif</span><span class="o">.</span><span class="n">as_string</span><span class="p">(</span><span class="n">_</span><span class="p">)</span>
<span class="go">&#39;two words&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cif</span><span class="o">.</span><span class="n">as_number</span><span class="p">(</span><span class="n">_</span><span class="p">)</span>
<span class="go">nan</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="reading-a-file">
<h2>Reading a file<a class="headerlink" href="#reading-a-file" title="Permalink to this headline">¶</a></h2>
<p>We have a few reading functions that read a file (or a string, or a stream)
and return a document (DOM) – an instance of class <code class="docutils literal notranslate"><span class="pre">Document</span></code>.</p>
<div class="section" id="id3">
<h3>C++<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<p>The reading functions are in the <code class="docutils literal notranslate"><span class="pre">gemmi::cif</span></code> namespace:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Document</span> <span class="n">read_file</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">filename</span><span class="p">)</span>
<span class="n">Document</span> <span class="n">read_memory</span><span class="p">(</span><span class="k">const</span> <span class="kt">char</span><span class="o">*</span> <span class="n">data</span><span class="p">,</span> <span class="k">const</span> <span class="kt">size_t</span> <span class="n">size</span><span class="p">,</span> <span class="k">const</span> <span class="kt">char</span><span class="o">*</span> <span class="n">name</span><span class="p">)</span>
<span class="n">Document</span> <span class="n">read_cstream</span><span class="p">(</span><span class="n">std</span><span class="o">::</span><span class="kt">FILE</span> <span class="o">*</span><span class="n">f</span><span class="p">,</span> <span class="kt">size_t</span> <span class="n">bufsize</span><span class="p">,</span> <span class="k">const</span> <span class="kt">char</span><span class="o">*</span> <span class="n">name</span><span class="p">)</span>
<span class="n">Document</span> <span class="n">read_istream</span><span class="p">(</span><span class="n">std</span><span class="o">::</span><span class="n">istream</span> <span class="o">&amp;</span><span class="n">is</span><span class="p">,</span> <span class="kt">size_t</span> <span class="n">bufsize</span><span class="p">,</span> <span class="k">const</span> <span class="kt">char</span><span class="o">*</span> <span class="n">name</span><span class="p">)</span>
</pre></div>
</div>
<p>Parameter <code class="docutils literal notranslate"><span class="pre">name</span></code> is used only when reporting errors.
Parameter <code class="docutils literal notranslate"><span class="pre">bufsize</span></code> determines the buffer size and only affects performance.
Regardless of the buffer size, the last two options are slower
than <code class="docutils literal notranslate"><span class="pre">read_file()</span></code> – they were not optimized for.</p>
<p>Additional header <code class="docutils literal notranslate"><span class="pre">&lt;gemmi/gz.hpp&gt;</span></code> is needed to transparently open
a gzipped file (by uncompressing it first into a memory buffer)
if the filename ends with <code class="docutils literal notranslate"><span class="pre">.gz</span></code>:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="c1">// in this and all the next examples: namespace cif = gemmi::cif;</span>
<span class="n">cif</span><span class="o">::</span><span class="n">Document</span> <span class="n">doc</span> <span class="o">=</span> <span class="n">cif</span><span class="o">::</span><span class="n">read</span><span class="p">(</span><span class="n">gemmi</span><span class="o">::</span><span class="n">MaybeGzipped</span><span class="p">(</span><span class="n">path</span><span class="p">));</span>
</pre></div>
</div>
<p>If you use it, you must also link the program with zlib. On Unix systems
it usually means adding <code class="docutils literal notranslate"><span class="pre">-lz</span></code> to the compiler invocation.</p>
<p>And if the <code class="docutils literal notranslate"><span class="pre">path</span></code> above is <code class="docutils literal notranslate"><span class="pre">-</span></code>, the standard input is read.</p>
</div>
<div class="section" id="id4">
<h3>Python<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">gemmi</span> <span class="kn">import</span> <span class="n">cif</span>

<span class="c1"># read and parse a CIF file</span>
<span class="n">doc</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">read_file</span><span class="p">(</span><span class="s1">&#39;components.cif&#39;</span><span class="p">)</span>

<span class="c1"># the same, but if the filename ends with .gz it is uncompressed on the fly</span>
<span class="n">doc</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;../tests/1pfe.cif.gz&#39;</span><span class="p">)</span>

<span class="c1"># read content of a CIF file from string</span>
<span class="n">doc</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">read_string</span><span class="p">(</span><span class="s1">&#39;data_this _is valid _cif content&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="low-level-functions">
<h3>Low-level functions<a class="headerlink" href="#low-level-functions" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">read_file()</span></code> call is equivalent to the following sequence:</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">path</span> <span class="o">=</span> <span class="s1">&#39;components.cif&#39;</span>
<span class="n">doc</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">Document</span><span class="p">()</span>
<span class="n">doc</span><span class="o">.</span><span class="n">source</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">doc</span><span class="o">.</span><span class="n">parse_file</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
<span class="n">doc</span><span class="o">.</span><span class="n">check_for_missing_values</span><span class="p">()</span>
<span class="n">doc</span><span class="o">.</span><span class="n">check_for_duplicates</span><span class="p">()</span>
</pre></div>
</div>
<p>The last two functions check for, respectively, missing values in tag-value
pairs and duplicated names.
It is possible to read erroneous CIF files by skipping these checks.</p>
<p>Analogically, function <code class="docutils literal notranslate"><span class="pre">read_string()</span></code> can be replaced by a similar
sequence with <code class="docutils literal notranslate"><span class="pre">Document.parse_string()</span></code> doing the main job:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">doc</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">Document</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">doc</span><span class="o">.</span><span class="n">parse_string</span><span class="p">(</span><span class="s1">&#39;data_this _tag_has_no_value</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">doc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">find_value</span><span class="p">(</span><span class="s1">&#39;_tag_has_no_value&#39;</span><span class="p">)</span>
<span class="go">&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">try</span><span class="p">:</span> <span class="n">doc</span><span class="o">.</span><span class="n">check_for_missing_values</span><span class="p">()</span>
<span class="gp">... </span><span class="k">except</span> <span class="ne">RuntimeError</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;missing value&#39;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">missing value</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="writing-a-file">
<h2>Writing a file<a class="headerlink" href="#writing-a-file" title="Permalink to this headline">¶</a></h2>
<p>Reading and writing a file does not preserve whitespaces.
Instead, we have a few choices for “styling” of the output:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Style::Simple</span></code> writes out the DOM structure adding blank lines between
mmCIF categories,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Style::NoBlankLines</span></code> does not add blank lines,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Style::PreferPairs</span></code> writes single-row loops as pairs,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Style::Pdbx</span></code> additionally puts <code class="docutils literal notranslate"><span class="pre">#</span></code> (empty comments) between categories,
mimicking the peculiar formatting of PDBx/mmCIF files in the official
wwPDB archive. It enables diff-ing original and modified files with
option <code class="docutils literal notranslate"><span class="pre">--ignore-space-change</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Style::Indent35</span></code> writes values from pairs from 35th column,</p></li>
</ul>
<div class="section" id="id5">
<h3>C++<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<p>The functions writing <code class="docutils literal notranslate"><span class="pre">cif::Document</span></code> to C++ stream
is in a separate header <code class="docutils literal notranslate"><span class="pre">gemmi/to_cif.hpp</span></code>:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="kt">void</span> <span class="n">write_cif_to_stream</span><span class="p">(</span><span class="n">std</span><span class="o">::</span><span class="n">ostream</span><span class="o">&amp;</span> <span class="n">os</span><span class="p">,</span> <span class="k">const</span> <span class="n">Document</span><span class="o">&amp;</span> <span class="n">doc</span><span class="p">,</span> <span class="n">Style</span> <span class="n">style</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id6">
<h3>Python<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<p>In Python, the function that writes the document to a file is a method
of the <code class="docutils literal notranslate"><span class="pre">Document</span></code> class:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">doc</span><span class="o">.</span><span class="n">write_file</span><span class="p">(</span><span class="s1">&#39;1pfe-modified.cif&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>It can take the style as optional, second argument:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">doc</span><span class="o">.</span><span class="n">write_file</span><span class="p">(</span><span class="s1">&#39;1pfe-modified.cif&#39;</span><span class="p">,</span> <span class="n">cif</span><span class="o">.</span><span class="n">Style</span><span class="o">.</span><span class="n">PreferPairs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">doc</span><span class="o">.</span><span class="n">write_file</span><span class="p">(</span><span class="s1">&#39;1pfe-styled.cif&#39;</span><span class="p">,</span> <span class="n">cif</span><span class="o">.</span><span class="n">Style</span><span class="o">.</span><span class="n">Pdbx</span><span class="p">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">Document</span></code> class also has a method <code class="docutils literal notranslate"><span class="pre">as_string()</span></code> which returns
the text that would be written by <code class="docutils literal notranslate"><span class="pre">write_file()</span></code>.</p>
</div>
</div>
<div class="section" id="document">
<h2>Document<a class="headerlink" href="#document" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">Document</span></code> is made of blocks with data. The blocks can be iterated over,
accessed by index or by name (each CIF block must have a unique name).</p>
<p>As it is common for cif files to contain only a single block,
gemmi has a method <code class="docutils literal notranslate"><span class="pre">sole_block()</span></code> that returns the first block
if the document has only one block; otherwise it throws an exception.</p>
<p>At last, is also has a member variable <code class="docutils literal notranslate"><span class="pre">source</span></code> that contains
the path of the file from which the document was read (if it was read
from a file).</p>
<div class="section" id="id7">
<h3>C++<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Document</span></code> has the two member variables:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">std</span><span class="o">::</span><span class="n">string</span> <span class="n">source</span><span class="p">;</span>  <span class="c1">// filename or the name passed to read_memory()</span>
<span class="n">std</span><span class="o">::</span><span class="n">vector</span><span class="o">&lt;</span><span class="n">Block</span><span class="o">&gt;</span> <span class="n">blocks</span><span class="p">;</span>
</pre></div>
</div>
<p>Each <code class="docutils literal notranslate"><span class="pre">Block</span></code> corresponds to a data block.
To access a block with known name use:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Block</span><span class="o">*</span> <span class="n">Document</span><span class="o">::</span><span class="n">find_block</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">name</span><span class="p">)</span>
</pre></div>
</div>
<p>To access the only block in the file you may use:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Block</span><span class="o">&amp;</span> <span class="n">Document</span><span class="o">::</span><span class="n">sole_block</span><span class="p">()</span>
</pre></div>
</div>
<p>A new <code class="docutils literal notranslate"><span class="pre">Document</span></code> instance can be created with default constructor.
To modify a document you need to access directly its member variables.
With one exception: when adding new blocks you can use a function that
additionally checks if the new name is unique:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Block</span><span class="o">&amp;</span> <span class="n">Document</span><span class="o">::</span><span class="n">add_new_block</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">name</span><span class="p">,</span> <span class="kt">int</span> <span class="n">pos</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id8">
<h3>Python<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Document</span></code> can be iterated, accessed by block index and by block name:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">doc</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">read_file</span><span class="p">(</span><span class="s2">&quot;components.cif&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">doc</span><span class="p">)</span>  
<span class="go">25219</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">doc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&lt;gemmi.cif.Block 000&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">doc</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&lt;gemmi.cif.Block ZZZ&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">doc</span><span class="p">[</span><span class="s1">&#39;MSE&#39;</span><span class="p">]</span>
<span class="go">&lt;gemmi.cif.Block MSE&gt;</span>
</pre></div>
</div>
<p>It has two other ways of accessing a block:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># The function block.find_block(name) is like block[name] ...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">doc</span><span class="o">.</span><span class="n">find_block</span><span class="p">(</span><span class="s1">&#39;MSE&#39;</span><span class="p">)</span>
<span class="go">&lt;gemmi.cif.Block MSE&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># ... except when the block is not found:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">doc</span><span class="o">.</span><span class="n">find_block</span><span class="p">(</span><span class="s1">&#39;no such thing&#39;</span><span class="p">)</span>  <span class="c1"># -&gt; None</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># doc[&#39;no such thing&#39;] # -&gt; KeyError</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Get the only block; throws exception if the document has more blocks.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cif</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;../tests/1pfe.cif.gz&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sole_block</span><span class="p">()</span>
<span class="go">&lt;gemmi.cif.Block 1PFE&gt;</span>
</pre></div>
</div>
<p>Blocks can be inserted (by default – appended after existing blocks)
using one of the two functions:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Document.add_new_block(name,</span> <span class="pre">pos=-1)</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Document.add_copied_block(block,</span> <span class="pre">pos=-1)</span></code></p></li>
</ul>
<p>As an example, here is how to start a new document:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">Document</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block_one</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">add_new_block</span><span class="p">(</span><span class="s1">&#39;block-one&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># populate block_one</span>
</pre></div>
</div>
<p>To delete a block use <code class="docutils literal notranslate"><span class="pre">Document.__delitem__</span></code> (for example: <code class="docutils literal notranslate"><span class="pre">del</span> <span class="pre">doc[1]</span></code>).</p>
<p>Document has also one property</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">doc</span><span class="o">.</span><span class="n">source</span>
<span class="go">&#39;components.cif&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="block">
<h2>Block<a class="headerlink" href="#block" title="Permalink to this headline">¶</a></h2>
<p>Each block has a name and a list of items.
Each item is one of:</p>
<ul class="simple">
<li><p>name-value pair (Pair),</p></li>
<li><p>table, a.k.a loop (Loop)</p></li>
<li><p>or save frame (Block – the same data structure as for block).</p></li>
</ul>
<div class="section" id="id9">
<h3>C++<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<p>Each block contains:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">std</span><span class="o">::</span><span class="n">string</span> <span class="n">name</span><span class="p">;</span>
<span class="n">std</span><span class="o">::</span><span class="n">vector</span><span class="o">&lt;</span><span class="n">Item</span><span class="o">&gt;</span> <span class="n">items</span><span class="p">;</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">Item</span></code> is implemented as an unrestricted (C++11) union
that holds one of Pair, Loop or Block.</p>
</div>
<div class="section" id="id10">
<h3>Python<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h3>
<p>Each block has a name:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">doc</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;../tests/1pfe.cif.gz&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">sole_block</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">name</span>
<span class="go">&#39;1PFE&#39;</span>
</pre></div>
</div>
<p>and a list of items (class Item):</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">block</span><span class="p">:</span>
<span class="gp">... </span>   <span class="k">if</span> <span class="n">item</span><span class="o">.</span><span class="n">line_number</span> <span class="o">&gt;</span> <span class="mi">1670</span><span class="p">:</span>
<span class="gp">... </span>       <span class="k">if</span> <span class="n">item</span><span class="o">.</span><span class="n">pair</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="gp">... </span>           <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;pair&#39;</span><span class="p">,</span> <span class="n">item</span><span class="o">.</span><span class="n">pair</span><span class="p">)</span>
<span class="gp">... </span>       <span class="k">elif</span> <span class="n">item</span><span class="o">.</span><span class="n">loop</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="gp">... </span>           <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;loop&#39;</span><span class="p">,</span> <span class="n">item</span><span class="o">.</span><span class="n">loop</span><span class="p">)</span>
<span class="gp">... </span>       <span class="k">elif</span> <span class="n">item</span><span class="o">.</span><span class="n">frame</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="gp">... </span>           <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;frame&#39;</span><span class="p">,</span> <span class="n">item</span><span class="o">.</span><span class="n">frame</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">pair [&#39;_ndb_struct_conf_na.entry_id&#39;, &#39;1PFE&#39;]</span>
<span class="go">pair [&#39;_ndb_struct_conf_na.feature&#39;, &quot;&#39;double helix&#39;&quot;]</span>
<span class="go">loop &lt;gemmi.cif.Loop 8 x 25&gt;</span>
<span class="go">loop &lt;gemmi.cif.Loop 5 x 43&gt;</span>
<span class="go">loop &lt;gemmi.cif.Loop 3 x 3&gt;</span>
<span class="go">loop &lt;gemmi.cif.Loop 83 x 10&gt;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="frame">
<h2>Frame<a class="headerlink" href="#frame" title="Permalink to this headline">¶</a></h2>
<p>(Very few people need it, skip this section.)</p>
<p>The <em>named save frames</em> (keyword <code class="docutils literal notranslate"><span class="pre">save_</span></code>) from the STAR specification
are used in CIF files only as sub-sections of a block.
The only place where they are enountered are mmCIF dictionaries.</p>
<p>The save-frame is stored as <code class="docutils literal notranslate"><span class="pre">Block</span></code> and can be accessed with:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Block</span><span class="o">*</span> <span class="n">Block</span><span class="o">::</span><span class="n">find_frame</span><span class="p">(</span><span class="n">std</span><span class="o">::</span><span class="n">string</span> <span class="n">name</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">frame</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find_frame</span><span class="p">(</span><span class="s1">&#39;my_frame&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Additionally, in C++ one may iterate over all Block’s items,
check each item type and handle all the save frames:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="p">(</span><span class="n">cif</span><span class="o">::</span><span class="n">Item</span><span class="o">&amp;</span> <span class="nl">item</span> <span class="p">:</span> <span class="n">block</span><span class="p">.</span><span class="n">items</span><span class="p">)</span>
  <span class="k">if</span> <span class="p">(</span><span class="n">item</span><span class="p">.</span><span class="n">type</span> <span class="o">==</span> <span class="n">cif</span><span class="o">::</span><span class="n">ItemType</span><span class="o">::</span><span class="n">Frame</span><span class="p">)</span>
    <span class="c1">// doing something with item.frame which is a (nested) Block</span>
    <span class="n">cif</span><span class="o">::</span><span class="n">Block</span><span class="o">&amp;</span> <span class="n">frame</span> <span class="o">=</span> <span class="n">item</span><span class="p">.</span><span class="n">frame</span><span class="p">;</span>
</pre></div>
</div>
</div>
<div class="section" id="pairs-and-loops">
<h2>Pairs and Loops<a class="headerlink" href="#pairs-and-loops" title="Permalink to this headline">¶</a></h2>
<p>The functions in this section can be considered low-level, because they are
specific to either name-value pairs or to loops.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Assuming what is a pair and what is in loop is a common source of bugs
when handling mmCIF files, so instead of these functions we recommend
using abstractions introduced in the next sections. For example,
when working with proteins one could assume that anisotropic ADP values
are in a loop, but wwPDB has entries with anisotropic ADP
for one atom only – as name-value pairs.
On the other hand, one could think that R-free is always given as
name-value, but in entries from joint X-ray and neutron refinement
it is in a loop.
Be careful.</p>
</div>
<p>The next sections introduce function that work with both pairs and loops.</p>
<div class="section" id="id11">
<h3>C++<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h3>
<p>Pair is simply defined as:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="k">using</span> <span class="n">Pair</span> <span class="o">=</span> <span class="n">std</span><span class="o">::</span><span class="n">array</span><span class="o">&lt;</span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="p">,</span> <span class="mi">2</span><span class="o">&gt;</span><span class="p">;</span>
</pre></div>
</div>
<p>A pair with a particular tag can be located using:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="k">const</span> <span class="n">Pair</span><span class="o">*</span> <span class="n">Block</span><span class="o">::</span><span class="n">find_pair</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">tag</span><span class="p">)</span> <span class="k">const</span>
</pre></div>
</div>
<p>or, if you want just the value:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">*</span> <span class="n">Block</span><span class="o">::</span><span class="n">find_value</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">tag</span><span class="p">)</span> <span class="k">const</span>
</pre></div>
</div>
<p>Both functions return <code class="docutils literal notranslate"><span class="pre">nullptr</span></code> if the tag is not found (and they
do not search in CIF loops).</p>
<p>To add a pair to the block, or modify an existing one, use:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="kt">void</span> <span class="n">Block</span><span class="o">::</span><span class="n">set_pair</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">tag</span><span class="p">,</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span> <span class="n">value</span><span class="p">)</span>
</pre></div>
</div>
<p>The value needs quoting, the passed argument needs to be already quoted
(you may pass <code class="docutils literal notranslate"><span class="pre">cif::quote(value)</span></code>).</p>
<hr class="docutils" />
<p>Loop is defined as:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="k">struct</span> <span class="n">Loop</span> <span class="p">{</span>
  <span class="n">std</span><span class="o">::</span><span class="n">vector</span><span class="o">&lt;</span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&gt;</span> <span class="n">tags</span><span class="p">;</span>
  <span class="n">std</span><span class="o">::</span><span class="n">vector</span><span class="o">&lt;</span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&gt;</span> <span class="n">values</span><span class="p">;</span>
  <span class="c1">// and a number of functions</span>
<span class="p">};</span>
</pre></div>
</div>
<p>To get values corresponding to a tag in a loop (table) you may use:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Column</span> <span class="n">Block</span><span class="o">::</span><span class="n">find_loop</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">tag</span><span class="p">)</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">struct</span> <span class="pre">Column</span></code>, which is documented further on, has method <code class="docutils literal notranslate"><span class="pre">get_loop()</span></code>
which gives access to <code class="docutils literal notranslate"><span class="pre">struct</span> <span class="pre">Loop</span></code>.</p>
<p>A new loop can be added using function:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Loop</span><span class="o">&amp;</span> <span class="n">Block</span><span class="o">::</span><span class="n">init_loop</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">prefix</span><span class="p">,</span> <span class="n">std</span><span class="o">::</span><span class="n">vector</span><span class="o">&lt;</span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&gt;</span> <span class="n">tags</span><span class="p">)</span>
</pre></div>
</div>
<p>Then it can be populated by either setting directly <code class="docutils literal notranslate"><span class="pre">tags</span></code> and <code class="docutils literal notranslate"><span class="pre">values</span></code>,
or my using Loop’s methods such as <code class="docutils literal notranslate"><span class="pre">add_row()</span></code> or <code class="docutils literal notranslate"><span class="pre">set_all_values()</span></code>.</p>
</div>
<div class="section" id="id12">
<h3>Python<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h3>
<p>Accessing name-value pairs:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># (1) tag and value</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">find_pair</span><span class="p">(</span><span class="s1">&#39;_cell.length_a&#39;</span><span class="p">)</span>
<span class="go">[&#39;_cell.length_a&#39;, &#39;39.374&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">find_pair</span><span class="p">(</span><span class="s1">&#39;_no_such_tag&#39;</span><span class="p">)</span>  <span class="c1"># return None</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c1"># (2) only value</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">find_value</span><span class="p">(</span><span class="s1">&#39;_cell.length_b&#39;</span><span class="p">)</span>
<span class="go">&#39;39.374&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">find_value</span><span class="p">(</span><span class="s1">&#39;_cell.no_such_tag&#39;</span><span class="p">)</span>  <span class="c1"># returns None</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c1"># (3) Item</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">item</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find_pair_item</span><span class="p">(</span><span class="s1">&#39;_cell.length_c&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">item</span><span class="o">.</span><span class="n">pair</span>
<span class="go">[&#39;_cell.length_c&#39;, &#39;79.734&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">item</span><span class="o">.</span><span class="n">line_number</span>
<span class="go">72</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">find_pair_item</span><span class="p">(</span><span class="s1">&#39;_nothing&#39;</span><span class="p">)</span>  <span class="c1"># return None</span>
</pre></div>
</div>
<p>To add a name-value pair, replacing current item if it exists,
use function <code class="docutils literal notranslate"><span class="pre">set_pair</span></code>:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">set_pair</span><span class="p">(</span><span class="s1">&#39;_year&#39;</span><span class="p">,</span> <span class="s1">&#39;2030&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>If the value needs quoting, it must be passed quoted:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">set_pair</span><span class="p">(</span><span class="s1">&#39;_title&#39;</span><span class="p">,</span> <span class="n">cif</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="s1">&#39;Goldilocks and the Three Bears&#39;</span><span class="p">))</span>
</pre></div>
</div>
<p>Now we can create a CIF file can from scratch:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">Document</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">add_new_block</span><span class="p">(</span><span class="s1">&#39;oak&#39;</span><span class="p">)</span>
<span class="go">&lt;gemmi.cif.Block oak&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span><span class="o">.</span><span class="n">set_pair</span><span class="p">(</span><span class="s1">&#39;_nut&#39;</span><span class="p">,</span> <span class="s1">&#39;acorn&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">as_string</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
<span class="go">data_oak</span>
<span class="go">_nut acorn</span>
</pre></div>
</div>
<hr class="docutils" />
<p>To access values in loop:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">find_loop</span><span class="p">(</span><span class="s1">&#39;_atom_type.symbol&#39;</span><span class="p">)</span>
<span class="go">&lt;gemmi.cif.Column _atom_type.symbol length 6&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">_</span><span class="p">)</span>
<span class="go">[&#39;C&#39;, &#39;CL&#39;, &#39;N&#39;, &#39;O&#39;, &#39;P&#39;, &#39;S&#39;]</span>
</pre></div>
</div>
<p>To add a row to an existing table (loop) use <code class="docutils literal notranslate"><span class="pre">add_row</span></code>:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">loop</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find_loop</span><span class="p">(</span><span class="s1">&#39;_atom_type.symbol&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">get_loop</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loop</span><span class="o">.</span><span class="n">add_row</span><span class="p">([</span><span class="s1">&#39;Au&#39;</span><span class="p">],</span> <span class="n">pos</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loop</span><span class="o">.</span><span class="n">add_row</span><span class="p">([</span><span class="s1">&#39;Zr&#39;</span><span class="p">])</span>  <span class="c1"># appends</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">find_loop</span><span class="p">(</span><span class="s1">&#39;_atom_type.symbol&#39;</span><span class="p">))</span>
<span class="go">[&#39;Au&#39;, &#39;C&#39;, &#39;CL&#39;, &#39;N&#39;, &#39;O&#39;, &#39;P&#39;, &#39;S&#39;, &#39;Zr&#39;]</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">set_all_values</span></code> sets all the data in a table. It takes as an argument
a list of lists of string. The lists of strings correspond to columns.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">loop</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find_loop</span><span class="p">(</span><span class="s1">&#39;_citation_author.citation_id&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">get_loop</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loop</span><span class="o">.</span><span class="n">tags</span>
<span class="go">[&#39;_citation_author.citation_id&#39;, &#39;_citation_author.name&#39;, &#39;_citation_author.ordinal&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loop</span><span class="o">.</span><span class="n">set_all_values</span><span class="p">([[</span><span class="s1">&#39;primary&#39;</span><span class="p">]</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="p">[</span><span class="n">cif</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="s1">&#39;Alice A.&#39;</span><span class="p">),</span> <span class="n">cif</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="s1">&#39;Bob B.&#39;</span><span class="p">)],</span> <span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;2&#39;</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">block</span><span class="o">.</span><span class="n">find_mmcif_category</span><span class="p">(</span><span class="s1">&#39;_citation_author.&#39;</span><span class="p">):</span>
<span class="gp">... </span>  <span class="nb">print</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">row</span><span class="p">))</span>
<span class="go">[&#39;primary&#39;, &quot;&#39;Alice A.&#39;&quot;, &#39;1&#39;]</span>
<span class="go">[&#39;primary&#39;, &quot;&#39;Bob B.&#39;&quot;, &#39;2&#39;]</span>
</pre></div>
</div>
<p>To add a new loop (replacing old one if it exists) use <code class="docutils literal notranslate"><span class="pre">init_loop</span></code>:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">loop</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">init_loop</span><span class="p">(</span><span class="s1">&#39;_ocean_&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># empty table is invalid in CIF, we need to add something</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loop</span><span class="o">.</span><span class="n">add_row</span><span class="p">([</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="n">cif</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="s1">&#39;Atlantic Ocean&#39;</span><span class="p">)])</span>
</pre></div>
</div>
<p>In the above example, if the block already has tags <code class="docutils literal notranslate"><span class="pre">_ocean_id</span></code>
and/or <code class="docutils literal notranslate"><span class="pre">_ocean_name</span></code> and</p>
<ul class="simple">
<li><p>if they are in a table: the table will be cleared and re-used,</p></li>
<li><p>if they are in name-value pairs: the pairs will be removed
and a table will be created at the position of the first pair.</p></li>
</ul>
<hr class="docutils" />
<p>To reorder items, use <code class="docutils literal notranslate"><span class="pre">block.move_item(old_pos,</span> <span class="pre">new_pos)</span></code>.
The current position of a tag can be obtained using <code class="docutils literal notranslate"><span class="pre">block.get_index(tag)</span></code>.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">get_index</span><span class="p">(</span><span class="s1">&#39;_entry.id&#39;</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">get_index</span><span class="p">(</span><span class="s1">&#39;_ocean_id&#39;</span><span class="p">)</span>
<span class="go">385</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">move_item</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># move first item to the end</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">get_index</span><span class="p">(</span><span class="s1">&#39;_entry.id&#39;</span><span class="p">)</span>
<span class="go">385</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">get_index</span><span class="p">(</span><span class="s1">&#39;_ocean_id&#39;</span><span class="p">)</span>
<span class="go">384</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">move_item</span><span class="p">(</span><span class="mi">385</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>  <span class="c1"># let&#39;s move it back (385 == -1 here)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="column">
<h2>Column<a class="headerlink" href="#column" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">Column</span></code> is a lightweight proxy class for working with both loop columns
and name-value pairs.</p>
<p>It was returned from <code class="docutils literal notranslate"><span class="pre">find_loop</span></code> above, but it is also returned from
a more general function <code class="docutils literal notranslate"><span class="pre">find_values()</span></code>, which searches for a given
tag in both loops and pairs.</p>
<div class="section" id="id13">
<h3>C++<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h3>
<p>The C++ signature of <code class="docutils literal notranslate"><span class="pre">find_values</span></code> is:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Column</span> <span class="n">Block</span><span class="o">::</span><span class="n">find_values</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">tag</span><span class="p">)</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">Column</span></code> has a few member functions:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="c1">// Number of rows in the loop. 0 means that the tag was not found.</span>
<span class="kt">int</span> <span class="nf">length</span><span class="p">()</span> <span class="k">const</span><span class="p">;</span>

<span class="c1">// Returns pointer to the column name in the DOM.</span>
<span class="c1">// The name is the same as argument to find_loop() or find_values().</span>
<span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">*</span> <span class="n">get_tag</span><span class="p">();</span>

<span class="c1">// Returns underlying Item (which contains either Pair or Loop).</span>
<span class="n">Item</span><span class="o">*</span> <span class="nf">item</span><span class="p">();</span>

<span class="c1">// Returns pointer to the DOM structure containing the whole table.</span>
<span class="n">Loop</span><span class="o">*</span> <span class="nf">get_loop</span><span class="p">()</span> <span class="k">const</span><span class="p">;</span>

<span class="c1">// Get raw value (no bounds checking).</span>
<span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="k">operator</span><span class="p">[](</span><span class="kt">int</span> <span class="n">n</span><span class="p">);</span>

<span class="c1">// Get raw value (after bounds checking).</span>
<span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">at</span><span class="p">(</span><span class="kt">int</span> <span class="n">n</span><span class="p">);</span>

<span class="c1">// Short-cut for cif::as_string(column.at(n)).</span>
<span class="n">std</span><span class="o">::</span><span class="n">string</span> <span class="n">str</span><span class="p">(</span><span class="kt">int</span> <span class="n">n</span><span class="p">)</span> <span class="k">const</span><span class="p">;</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">Column</span></code> also provides support for C++11 range-based <code class="docutils literal notranslate"><span class="pre">for</span></code>:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="c1">// mmCIF _chem_comp_atom is usually a table, but not always</span>
<span class="k">for</span> <span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span> <span class="o">&amp;</span><span class="nl">s</span> <span class="p">:</span> <span class="n">block</span><span class="p">.</span><span class="n">find_values</span><span class="p">(</span><span class="s">&quot;_chem_comp_atom.type_symbol&quot;</span><span class="p">))</span>
  <span class="n">std</span><span class="o">::</span><span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="n">s</span> <span class="o">&lt;&lt;</span> <span class="n">std</span><span class="o">::</span><span class="n">endl</span><span class="p">;</span>
</pre></div>
</div>
<p>If the column represents a name-value pair,
<code class="docutils literal notranslate"><span class="pre">Column::get_loop()</span></code> return <code class="docutils literal notranslate"><span class="pre">nullptr</span></code>
(and <code class="docutils literal notranslate"><span class="pre">Column::get_length()</span></code> returns 1).</p>
</div>
<div class="section" id="id14">
<h3>Python<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h3>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">find_values</span><span class="p">(</span><span class="s1">&#39;_cell.length_a&#39;</span><span class="p">)</span>  <span class="c1"># name-value pair</span>
<span class="go">&lt;gemmi.cif.Column _cell.length_a length 1&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">find_values</span><span class="p">(</span><span class="s1">&#39;_atom_site.Cartn_x&#39;</span><span class="p">)</span>  <span class="c1"># column in a loop</span>
<span class="go">&lt;gemmi.cif.Column _atom_site.Cartn_x length 342&gt;</span>
</pre></div>
</div>
<p>Column’s special method <code class="docutils literal notranslate"><span class="pre">__bool__</span></code> tells if the tag was found.
<code class="docutils literal notranslate"><span class="pre">__len__</span></code> returns the number of corresponding values.
<code class="docutils literal notranslate"><span class="pre">__iter__</span></code>,  <code class="docutils literal notranslate"><span class="pre">__getitem__</span></code> and <code class="docutils literal notranslate"><span class="pre">__setitem__</span></code>
get or set a raw string (i.e. string with quotes, if applicable).</p>
<p>As an example, let us shift all the atoms by 1A in the x direction:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">col_x</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find_values</span><span class="p">(</span><span class="s1">&#39;_atom_site.Cartn_x&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">col_x</span><span class="p">):</span>
<span class="gp">... </span>  <span class="n">col_x</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<p>To get the actual string content (without quotes) one may use
the method <code class="docutils literal notranslate"><span class="pre">str</span></code>:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">column</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find_values</span><span class="p">(</span><span class="s1">&#39;_chem_comp.formula&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">column</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>
<span class="go">&quot;&#39;H2 O&#39;&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">column</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span>  <span class="c1"># short-cut for cif.as_string(column[7])</span>
<span class="go">&#39;H2 O&#39;</span>
</pre></div>
</div>
<p>If the tag is found in a loop, method <code class="docutils literal notranslate"><span class="pre">get_loop</span></code> returns a reference
to this <code class="docutils literal notranslate"><span class="pre">Loop</span></code> in the DOM. Otherwise it returns <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">column</span><span class="o">.</span><span class="n">get_loop</span><span class="p">()</span>
<span class="go">&lt;gemmi.cif.Loop 12 x 7&gt;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="table">
<h2>Table<a class="headerlink" href="#table" title="Permalink to this headline">¶</a></h2>
<p>Usually we want to access multiple columns at once,
so the library has another abstraction (<code class="docutils literal notranslate"><span class="pre">Table</span></code>)
that can be used with multiple tags.</p>
<p><code class="docutils literal notranslate"><span class="pre">Table</span></code> is returned by <code class="docutils literal notranslate"><span class="pre">Block.find()</span></code>.
Like column, it is a lightweight, iterable view of the data,
but it is for querying multiple related tags at the same time.</p>
<p>The first form of <code class="docutils literal notranslate"><span class="pre">find()</span></code> takes a list of tags:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Table</span> <span class="n">Block</span><span class="o">::</span><span class="n">find</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">vector</span><span class="o">&lt;</span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&gt;&amp;</span> <span class="n">tags</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">find</span><span class="p">([</span><span class="s1">&#39;_entity_poly_seq.entity_id&#39;</span><span class="p">,</span> <span class="s1">&#39;_entity_poly_seq.num&#39;</span><span class="p">,</span> <span class="s1">&#39;_entity_poly_seq.mon_id&#39;</span><span class="p">])</span>
<span class="go">&lt;gemmi.cif.Table 18 x 3&gt;</span>
</pre></div>
</div>
<p>Since tags in one loop tend to have a common prefix (category name),
the library provides also a second form that takes the common prefix
as the first argument:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Table</span> <span class="n">Block</span><span class="o">::</span><span class="n">find</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">prefix</span><span class="p">,</span> <span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">vector</span><span class="o">&lt;</span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&gt;&amp;</span> <span class="n">tags</span><span class="p">)</span>

<span class="c1">// These two calls are equivalent:</span>

<span class="n">block</span><span class="p">.</span><span class="n">find</span><span class="p">({</span><span class="s">&quot;_entity_poly_seq.entity_id&quot;</span><span class="p">,</span> <span class="s">&quot;_entity_poly_seq.num&quot;</span><span class="p">,</span> <span class="s">&quot;_entity_poly_seq.mon_id&quot;</span><span class="p">});</span>
<span class="n">block</span><span class="p">.</span><span class="n">find</span><span class="p">(</span><span class="s">&quot;_entity_poly_seq.&quot;</span><span class="p">,</span> <span class="p">{</span><span class="s">&quot;entity_id&quot;</span><span class="p">,</span> <span class="s">&quot;num&quot;</span><span class="p">,</span> <span class="s">&quot;mon_id&quot;</span><span class="p">});</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;_entity_poly_seq.&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;entity_id&#39;</span><span class="p">,</span> <span class="s1">&#39;num&#39;</span><span class="p">,</span> <span class="s1">&#39;mon_id&#39;</span><span class="p">])</span>
<span class="go">&lt;gemmi.cif.Table 18 x 3&gt;</span>
</pre></div>
</div>
<p>Note that <code class="docutils literal notranslate"><span class="pre">find</span></code> is not aware of dictionaries and categories,
therefore the category name should end with a separator
(dot for mmCIF files, as shown above).</p>
<p>In the example above, all the tags are required. If one of them is absent,
the returned <code class="docutils literal notranslate"><span class="pre">Table</span></code> is empty. Tags (all except the first one) can be marked
as <em>optional</em> by adding prefix <code class="docutils literal notranslate"><span class="pre">?</span></code>:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Table</span> <span class="n">table</span> <span class="o">=</span> <span class="n">block</span><span class="p">.</span><span class="n">find</span><span class="p">({</span><span class="s">&quot;_required_tag&quot;</span><span class="p">,</span> <span class="s">&quot;?_optional_tag&quot;</span><span class="p">})</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">table</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find</span><span class="p">([</span><span class="s1">&#39;_required_tag&#39;</span><span class="p">,</span> <span class="s1">&#39;?_optional_tag&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p>In such case the returned table may contain either one or two columns.
Before accessing column corresponding to an optional tag one must check
if the column exists with <code class="docutils literal notranslate"><span class="pre">Table::has_column()</span></code> (or, alternatively,
with equivalent function <code class="docutils literal notranslate"><span class="pre">Table::Row::has()</span></code> which will be introduced
later):</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="kt">bool</span> <span class="n">Table</span><span class="o">::</span><span class="n">has_column</span><span class="p">(</span><span class="kt">int</span> <span class="n">n</span><span class="p">)</span> <span class="k">const</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;_entity_poly_seq.&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;entity_id&#39;</span><span class="p">,</span> <span class="s1">&#39;?num&#39;</span><span class="p">,</span> <span class="s1">&#39;?bleh&#39;</span><span class="p">])</span>
<span class="go">&lt;gemmi.cif.Table 18 x 3&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span><span class="o">.</span><span class="n">has_column</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="n">_</span><span class="o">.</span><span class="n">has_column</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">_</span><span class="o">.</span><span class="n">has_column</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="go">(True, True, False)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">Table</span></code> has functions to check its shape:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="kt">bool</span> <span class="nf">ok</span><span class="p">()</span> <span class="k">const</span><span class="p">;</span>  <span class="c1">// true if the table is not empty</span>
<span class="kt">size_t</span> <span class="nf">width</span><span class="p">()</span> <span class="k">const</span><span class="p">;</span>  <span class="c1">// number of columns</span>
<span class="kt">size_t</span> <span class="nf">length</span><span class="p">()</span> <span class="k">const</span><span class="p">;</span>  <span class="c1">// number of rows</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">table</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;_entity_poly_seq.&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;entity_id&#39;</span><span class="p">,</span> <span class="s1">&#39;num&#39;</span><span class="p">,</span> <span class="s1">&#39;mon_id&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># instead of ok() in Python we use __bool__()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="n">table</span><span class="p">,</span> <span class="s2">&quot;table.__bool__() is expected to return True&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">table</span><span class="o">.</span><span class="n">width</span><span class="p">()</span>  <span class="c1"># number of columns</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">table</span><span class="p">)</span>  <span class="c1"># number of rows</span>
<span class="go">18</span>
</pre></div>
</div>
<p>If Table’s data is in Loop, the Loop class can be accessed using:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Loop</span><span class="o">*</span> <span class="nf">get_loop</span><span class="p">();</span>  <span class="c1">// nullptr for tag-value pairs</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">table</span><span class="o">.</span><span class="n">loop</span>     <span class="c1"># None for tag-value pairs</span>
<span class="go">&lt;gemmi.cif.Loop 18 x 4&gt;</span>
</pre></div>
</div>
<p>If a prefix was specified when calling find, the prefix length is stored
and the prefix can be retrived:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="kt">size_t</span> <span class="n">prefix_length</span><span class="p">;</span>
<span class="n">std</span><span class="o">::</span><span class="n">string</span> <span class="n">get_prefix</span><span class="p">()</span> <span class="k">const</span><span class="p">;</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">table</span><span class="o">.</span><span class="n">prefix_length</span>
<span class="go">17</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">table</span><span class="o">.</span><span class="n">get_prefix</span><span class="p">()</span>
<span class="go">&#39;_entity_poly_seq.&#39;</span>
</pre></div>
</div>
<p>Table also has function <code class="docutils literal notranslate"><span class="pre">erase()</span></code> that deletes all tags and data
associated with the table.
See an example in the <a class="reference internal" href="#ccd-example"><span class="std std-ref">CCD section</span></a> below.</p>
<div class="section" id="row-wise-access">
<h3>Row-wise access<a class="headerlink" href="#row-wise-access" title="Permalink to this headline">¶</a></h3>
<p>Most importantly, <code class="docutils literal notranslate"><span class="pre">Table</span></code> provides access to rows (<code class="docutils literal notranslate"><span class="pre">Table::Row</span></code>)
that in turn provide access to value strings:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Row</span> <span class="n">Table</span><span class="o">::</span><span class="k">operator</span><span class="p">[](</span><span class="kt">int</span> <span class="n">n</span><span class="p">)</span>  <span class="c1">// access Row</span>
<span class="n">Row</span> <span class="n">Table</span><span class="o">::</span><span class="n">at</span><span class="p">(</span><span class="kt">int</span> <span class="n">n</span><span class="p">)</span>          <span class="c1">// the same but with bounds checking</span>
<span class="c1">// and also begin(), end() and iterator that enable range-for.</span>

<span class="c1">// Returns the first row that has the specified string in the first column.</span>
<span class="n">Row</span> <span class="n">Table</span><span class="o">::</span><span class="n">find_row</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">s</span><span class="p">)</span>

<span class="c1">// Makes sure that the table has only one row and returns it.</span>
<span class="n">Row</span> <span class="n">Table</span><span class="o">::</span><span class="n">one</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">table</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&lt;gemmi.cif.Table.Row: 1 1 DG&gt;</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c1"># and of course it&#39;s iterable</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">table</span><span class="p">:</span> <span class="k">pass</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Returns the first row that has the specified string in the first column.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">table</span><span class="o">.</span><span class="n">find_row</span><span class="p">(</span><span class="s1">&#39;2&#39;</span><span class="p">)</span>
<span class="go">&lt;gemmi.cif.Table.Row: 2 1 DSN&gt;</span>
</pre></div>
</div>
<p>as well as to the tags:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Row</span> <span class="nf">tags</span><span class="p">();</span>  <span class="c1">// pseudo-row that contains tags</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">table</span><span class="o">.</span><span class="n">tags</span>
<span class="go">&lt;gemmi.cif.Table.Row: _entity_poly_seq.entity_id _entity_poly_seq.num _entity_poly_seq.mon_id&gt;</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">Table::Row</span></code> has functions for accessing the values:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="c1">// Get raw value.</span>
<span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">Table</span><span class="o">::</span><span class="n">Row</span><span class="o">::</span><span class="k">operator</span><span class="p">[](</span><span class="kt">int</span> <span class="n">n</span><span class="p">)</span>  <span class="c1">// no bounds checking</span>
<span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">Table</span><span class="o">::</span><span class="n">Row</span><span class="o">::</span><span class="n">at</span><span class="p">(</span><span class="kt">int</span> <span class="n">n</span><span class="p">)</span>          <span class="c1">// with bounds checking</span>
<span class="c1">// and also begin(), end(), iterator, const_supports iterators.</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">table</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
<span class="go">DG,DC,DG,DT,DA,DC,DG,DC,DSN,ALA,N2C,NCY,MVA,DSN,ALA,NCY,N2C,MVA,</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">row</span> <span class="o">=</span> <span class="n">table</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">row</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
<span class="go">2,2,ALA,</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="go">&#39;ALA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">row</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>  <span class="c1"># the same</span>
<span class="go">&#39;ALA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="c1"># the same, but returns None instead of IndexError</span>
<span class="go">&#39;ALA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;mon_id&#39;</span><span class="p">]</span>  <span class="c1"># the same, but slightly slower than numeric index</span>
<span class="go">&#39;ALA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;_entity_poly_seq.mon_id&#39;</span><span class="p">]</span>  <span class="c1"># the same</span>
<span class="go">&#39;ALA&#39;</span>
</pre></div>
</div>
<p>and a few convenience functions, including:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="kt">size_t</span> <span class="n">Table</span><span class="o">::</span><span class="n">Row</span><span class="o">::</span><span class="n">size</span><span class="p">()</span> <span class="k">const</span>           <span class="c1">// the width of the table</span>
<span class="n">std</span><span class="o">::</span><span class="n">string</span> <span class="n">Table</span><span class="o">::</span><span class="n">Row</span><span class="o">::</span><span class="n">str</span><span class="p">(</span><span class="kt">int</span> <span class="n">n</span><span class="p">)</span> <span class="k">const</span>  <span class="c1">// short-cut for cif::as_string(row.at(n))</span>
<span class="kt">bool</span> <span class="n">Table</span><span class="o">::</span><span class="n">Row</span><span class="o">::</span><span class="n">has</span><span class="p">(</span><span class="kt">int</span> <span class="n">n</span><span class="p">)</span> <span class="k">const</span>         <span class="c1">// the same as Table::has_column(n)</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>    <span class="c1"># the same as table.width()</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">row</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>  <span class="c1"># if the value is in quotes, it gets un-quoted</span>
<span class="go">&#39;ALA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">row</span><span class="o">.</span><span class="n">has</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>  <span class="c1"># the same as table.has_column(2)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>The tags and value can be modified. As an example, let us swap two tag names
(these two tend to have identical values, so no one will notice):</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span>  <span class="n">cif</span><span class="o">::</span><span class="n">Table</span><span class="o">::</span><span class="n">Row</span> <span class="n">tags</span> <span class="o">=</span> <span class="n">block</span><span class="p">.</span><span class="n">find</span><span class="p">(</span><span class="s">&quot;_atom_site.&quot;</span><span class="p">,</span> <span class="p">{</span><span class="s">&quot;label_atom_id&quot;</span><span class="p">,</span> <span class="s">&quot;auth_atom_id&quot;</span><span class="p">}).</span><span class="n">tags</span><span class="p">();</span>
  <span class="n">std</span><span class="o">::</span><span class="n">swap</span><span class="p">(</span><span class="n">tags</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">tags</span><span class="p">[</span><span class="mi">1</span><span class="p">]);</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tags</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;_atom_site.&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;label_atom_id&#39;</span><span class="p">,</span> <span class="s1">&#39;auth_atom_id&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">tags</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tags</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">tags</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">tags</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">tags</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="section" id="column-wise-access">
<h3>Column-wise access<a class="headerlink" href="#column-wise-access" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Table</span></code> gives also access to columns, represented by the previously
introduced <code class="docutils literal notranslate"><span class="pre">Column</span></code>:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Column</span> <span class="n">Table</span><span class="o">::</span><span class="n">column</span><span class="p">(</span><span class="kt">int</span> <span class="n">index</span><span class="p">)</span>

<span class="c1">// alternatively, specify tag name</span>
<span class="n">Column</span> <span class="n">Table</span><span class="o">::</span><span class="n">find_column</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">tag</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">table</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="go">&lt;gemmi.cif.Column _entity_poly_seq.entity_id length 18&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">table</span><span class="o">.</span><span class="n">find_column</span><span class="p">(</span><span class="s1">&#39;_entity_poly_seq.mon_id&#39;</span><span class="p">)</span>
<span class="go">&lt;gemmi.cif.Column _entity_poly_seq.mon_id length 18&gt;</span>
</pre></div>
</div>
<p>If the table is created in a function that uses prefix,
the prefix can be omitted in <code class="docutils literal notranslate"><span class="pre">find_column</span></code>:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Table</span> <span class="n">t</span> <span class="o">=</span> <span class="n">block</span><span class="p">.</span><span class="n">find</span><span class="p">(</span><span class="s">&quot;_entity_poly_seq.&quot;</span><span class="p">,</span> <span class="p">{</span><span class="s">&quot;entity_id&quot;</span><span class="p">,</span> <span class="s">&quot;num&quot;</span><span class="p">,</span> <span class="s">&quot;mon_id&quot;</span><span class="p">});</span>
<span class="n">Column</span> <span class="n">col</span> <span class="o">=</span> <span class="n">t</span><span class="p">.</span><span class="n">find_column</span><span class="p">(</span><span class="mi">2</span><span class="p">);</span>
<span class="c1">// is equivalent to</span>
<span class="n">Column</span> <span class="n">col</span> <span class="o">=</span> <span class="n">t</span><span class="p">.</span><span class="n">find_column</span><span class="p">(</span><span class="s">&quot;_entity_poly_seq.mon_id&quot;</span><span class="p">);</span>
<span class="c1">// is equivalent to</span>
<span class="n">Column</span> <span class="n">col</span> <span class="o">=</span> <span class="n">t</span><span class="p">.</span><span class="n">find_column</span><span class="p">(</span><span class="s">&quot;mon_id&quot;</span><span class="p">);</span>
</pre></div>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">table</span><span class="o">.</span><span class="n">find_column</span><span class="p">(</span><span class="s1">&#39;mon_id&#39;</span><span class="p">)</span>
<span class="go">&lt;gemmi.cif.Column _entity_poly_seq.mon_id length 18&gt;</span>
</pre></div>
</div>
<p>C++ note:
both <code class="docutils literal notranslate"><span class="pre">Column</span></code> and <code class="docutils literal notranslate"><span class="pre">Table::Row</span></code> have functions <code class="docutils literal notranslate"><span class="pre">begin()</span></code> and <code class="docutils literal notranslate"><span class="pre">end()</span></code>
in const and non-const variants, returning <code class="docutils literal notranslate"><span class="pre">iterator</span></code> and
<code class="docutils literal notranslate"><span class="pre">const_iterator</span></code> types, respectively. These types satisfy requirements
of the BidirectionalIterator concept.
Conversely, the iterator over the rows of <code class="docutils literal notranslate"><span class="pre">Table</span></code> is a minimalistic
structure – just enough get the range-for work.</p>
</div>
<div class="section" id="mmcif-categories">
<h3>mmCIF categories<a class="headerlink" href="#mmcif-categories" title="Permalink to this headline">¶</a></h3>
<p>mmCIF files group data into categories. All mmCIF tags have a dot
(e.g. <code class="docutils literal notranslate"><span class="pre">_entry.id</span></code>) and the category name is the part before the dot.</p>
<div class="section" id="id15">
<h4>C++<a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h4>
<p>We have two functions to work with categories.
One returns a list of all categories in the block:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">std</span><span class="o">::</span><span class="n">vector</span><span class="o">&lt;</span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&gt;</span> <span class="n">Block</span><span class="o">::</span><span class="n">get_mmcif_category_names</span><span class="p">()</span> <span class="k">const</span>
</pre></div>
</div>
<p>The other returns a <code class="docutils literal notranslate"><span class="pre">Table</span></code> with all tags (and values) belonging to
the specified category:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="n">Table</span> <span class="n">Block</span><span class="o">::</span><span class="n">find_mmcif_category</span><span class="p">(</span><span class="n">std</span><span class="o">::</span><span class="n">string</span> <span class="n">cat</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id16">
<h4>Python<a class="headerlink" href="#id16" title="Permalink to this headline">¶</a></h4>
<p>Python bindings have the same two functions:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">get_mmcif_category_names</span><span class="p">()[:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">[&#39;_entry.&#39;, &#39;_audit_conform.&#39;, &#39;_database_2.&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">find_mmcif_category</span><span class="p">(</span><span class="s1">&#39;_entry.&#39;</span><span class="p">)</span>
<span class="go">&lt;gemmi.cif.Table 1 x 1&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span><span class="o">.</span><span class="n">tags</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">_</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="go">(&#39;_entry.id&#39;, &#39;1PFE&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cat</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find_mmcif_category</span><span class="p">(</span><span class="s1">&#39;_database_2.&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cat</span>
<span class="go">&lt;gemmi.cif.Table 4 x 2&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">cat</span><span class="o">.</span><span class="n">tags</span><span class="p">)</span>
<span class="go">[&#39;_database_2.database_id&#39;, &#39;_database_2.database_code&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">cat</span><span class="p">:</span>
<span class="gp">... </span>  <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1">: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">row</span><span class="p">))</span>
<span class="go">PDB: 1PFE</span>
<span class="go">NDB: DD0057</span>
<span class="go">RCSB: RCSB019291</span>
<span class="go">WWPDB: D_1000019291</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cat</span><span class="p">[</span><span class="mi">3</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;D_1000019291&#39;</span>
</pre></div>
</div>
<p>Additionally, two Python-specific functions: <code class="docutils literal notranslate"><span class="pre">get_mmcif_category</span></code>
and <code class="docutils literal notranslate"><span class="pre">set_mmcif_category</span></code> translate between an mmCIF category and
Python dictionary:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">get_mmcif_category</span><span class="p">(</span><span class="s1">&#39;_entry.&#39;</span><span class="p">)</span>
<span class="go">{&#39;id&#39;: [&#39;1PFE&#39;]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">get_mmcif_category</span><span class="p">(</span><span class="s1">&#39;_database_2.&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<span class="go">[&#39;database_code&#39;, &#39;database_id&#39;]</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">?</span></code> and <code class="docutils literal notranslate"><span class="pre">.</span></code> are translated to <code class="docutils literal notranslate"><span class="pre">None</span></code> and <code class="docutils literal notranslate"><span class="pre">False</span></code>.
To disable this translation and get “raw” strings, add argument <code class="docutils literal notranslate"><span class="pre">raw=True</span></code>.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">default</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">get_mmcif_category</span><span class="p">(</span><span class="s1">&#39;_software&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">raw</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">get_mmcif_category</span><span class="p">(</span><span class="s1">&#39;_software&#39;</span><span class="p">,</span> <span class="n">raw</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;version&#39;</span><span class="p">,</span> <span class="s1">&#39;citation_id&#39;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">default</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">raw</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
<span class="gp">...</span>
<span class="go">EPMR EPMR</span>
<span class="go">False .</span>
<span class="go">None ?</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">set_mmcif_category()</span></code> takes a category name and a dictionary (that
maps tags to lists) and creates or replaces the corresponding category:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ocean_data</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;Atlantic&#39;</span><span class="p">,</span> <span class="s1">&#39;Pacific&#39;</span><span class="p">,</span> <span class="s1">&#39;Indian&#39;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">set_mmcif_category</span><span class="p">(</span><span class="s1">&#39;_ocean&#39;</span><span class="p">,</span> <span class="n">ocean_data</span><span class="p">)</span>
</pre></div>
</div>
<p>It also takes the optional <code class="docutils literal notranslate"><span class="pre">raw</span></code> parameter (default: <code class="docutils literal notranslate"><span class="pre">False</span></code>).
Specify <code class="docutils literal notranslate"><span class="pre">raw=True</span></code> if the values are already quoted, otherwise
they will be quoted twice.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">set_mmcif_category</span><span class="p">(</span><span class="s1">&#39;_raw&#39;</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;a&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;?&#39;</span><span class="p">,</span> <span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="s2">&quot;&#39;a b&#39;&quot;</span><span class="p">]},</span> <span class="n">raw</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">find_values</span><span class="p">(</span><span class="s1">&#39;_raw.a&#39;</span><span class="p">))</span>
<span class="go">[&#39;?&#39;, &#39;.&#39;, &quot;&#39;a b&#39;&quot;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">set_mmcif_category</span><span class="p">(</span><span class="s1">&#39;_nop&#39;</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;a&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;?&#39;</span><span class="p">,</span> <span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="s2">&quot;&#39;a b&#39;&quot;</span><span class="p">]},</span> <span class="n">raw</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">find_values</span><span class="p">(</span><span class="s1">&#39;_nop.a&#39;</span><span class="p">))</span>
<span class="go">[&quot;&#39;?&#39;&quot;, &quot;&#39;.&#39;&quot;, &#39;&quot;\&#39;a b\&#39;&quot;&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span><span class="o">.</span><span class="n">set_mmcif_category</span><span class="p">(</span><span class="s1">&#39;_ok&#39;</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;a&#39;</span><span class="p">:</span> <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="s1">&#39;a b&#39;</span><span class="p">]},</span> <span class="n">raw</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">find_values</span><span class="p">(</span><span class="s1">&#39;_ok.a&#39;</span><span class="p">))</span>
<span class="go">[&#39;?&#39;, &#39;.&#39;, &quot;&#39;a b&#39;&quot;]</span>
</pre></div>
</div>
<p>Finally, we have a variant of <code class="docutils literal notranslate"><span class="pre">init_loop</span></code> for working with mmCIF categories:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">loop</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">init_mmcif_loop</span><span class="p">(</span><span class="s1">&#39;_ocean.&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loop</span><span class="o">.</span><span class="n">add_row</span><span class="p">([</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;Atlantic&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p>The subtle difference from generic <code class="docutils literal notranslate"><span class="pre">init_loop</span></code> is that if the block
has other name-value pairs in the same category (say, <code class="docutils literal notranslate"><span class="pre">_ocean.depth</span> <span class="pre">8.5</span></code>)
<code class="docutils literal notranslate"><span class="pre">init_loop</span></code> leaves them untouched, but <code class="docutils literal notranslate"><span class="pre">init_mmcif_loop</span></code> removes them.
Additionally, like in other <code class="docutils literal notranslate"><span class="pre">_mmcif_</span></code> functions, the trailing dot
in the category name may be omitted (but the leading underscore is required).</p>
</div>
</div>
</div>
<div class="section" id="json">
<h2>JSON<a class="headerlink" href="#json" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">cif::Document</span></code> can be stored in a JSON format, and it can be read from
JSON file. This is in general true about CIF files - their content can
be converted to JSON and back. In gemmi we have a number of options to
customize the translation. In particular, both mmCIF and CIF-JSON flavours
are supported.
More details about the flavours are given in the description of
<a class="reference internal" href="utils.html#json"><span class="std std-ref">gemmi convert</span></a>.</p>
<div class="section" id="id17">
<h3>C++<a class="headerlink" href="#id17" title="Permalink to this headline">¶</a></h3>
<p>Header <code class="docutils literal notranslate"><span class="pre">gemmi/to_json.hpp</span></code> provides code for serializing
<code class="docutils literal notranslate"><span class="pre">cif::Document</span></code> as JSON.</p>
<p>Such JSON files can be read back into the <code class="docutils literal notranslate"><span class="pre">cif::Document</span></code> structure
using function from <code class="docutils literal notranslate"><span class="pre">gemmi/json.hpp</span></code>.</p>
</div>
<div class="section" id="id18">
<h3>Python<a class="headerlink" href="#id18" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Document.as_json()</span></code> returns the document serialized in JSON string.
To output mmJSON add argument <code class="docutils literal notranslate"><span class="pre">mmjson=True</span></code>.</p>
<p>mmJSON (possibly gzipped) can be read using function <code class="docutils literal notranslate"><span class="pre">cif.read_mmjson()</span></code>.
In addition, the function <code class="docutils literal notranslate"><span class="pre">cif.read()</span></code> will also read mmJSON format
if the file name ends with <code class="docutils literal notranslate"><span class="pre">.json</span></code> or <code class="docutils literal notranslate"><span class="pre">.json.gz</span></code>.</p>
</div>
</div>
<div class="section" id="design-choices">
<h2>Design choices<a class="headerlink" href="#design-choices" title="Permalink to this headline">¶</a></h2>
<div class="section" id="parser">
<h3>Parser<a class="headerlink" href="#parser" title="Permalink to this headline">¶</a></h3>
<p>Parsing formal languages is a well-researched topic in computer science.
The first versions of lex and yacc - popular tools that generate lexical
analyzers and parsers - were written in 1970’s. Today many tools exist
to translate grammar rules into C/C++ code.
On the other hand many compilers and high-profile tools use hand-coded
parsers - as it is more flexible.</p>
<p>Looking at other STAR and CIF parsers – some use parser generators
(COD::CIF::Parser and starlib2 use yacc/bison, iotbx.cif uses Antlr3),
others are hand-coded.</p>
<p>I had experience with flex/bison and Boost.Spirit
(and I wanted to try also Lemon and re2c)
but I decided to use PEGTL for this task (and I’m very happy with this choice).
I was convinced by the
<a class="reference external" href="https://github.com/taocpp/json">TAOC++ JSON</a>
parser that is based on PEGTL and has a good balance of simplicity
and performance.</p>
<p><a class="reference external" href="https://github.com/taocpp/PEGTL/">PEGTL</a> is a C++ library (not a generator)
for creating PEG parsers.
PEG stands for Parsing Expression Grammar – a simpler approach than
the traditional Context Free Grammar.</p>
<p>As a result, our parser depends on a third-party (header-only) library,
but the parser itself is pretty simple.</p>
</div>
<div class="section" id="data-structures">
<h3>Data structures<a class="headerlink" href="#data-structures" title="Permalink to this headline">¶</a></h3>
<p>The next thing is how we store the data read from file.
We decided to rely on the C++ standard library where we can.</p>
<p>Generally, storage of such data involves (in C++ terms) some containers
and a union/variant type for storing values of mixed types.</p>
<p>We use primarily <code class="docutils literal notranslate"><span class="pre">std::vector</span></code> as a container.</p>
<p>A custom structure <code class="docutils literal notranslate"><span class="pre">Item</span></code> with (unrestricted) union is used as
a variant-like class.</p>
<p>Strings are stored in <code class="docutils literal notranslate"><span class="pre">std::string</span></code> and it is fast enough.
Mainstream C++ standard libraries have short string optimization (SSO)
for up to 15 or 22 characters, which covers most of the values in mmCIF files.</p>
</div>
</div>
<div class="section" id="performance">
<h2>Performance<a class="headerlink" href="#performance" title="Permalink to this headline">¶</a></h2>
<p>Gemmi has <a class="reference external" href="https://github.com/project-gemmi/mmcif-benchmark">the fastest</a>
open-source CIF parser (at least in the hands of the author).
While further improvement would be possible (some JSON parsers are
<a class="reference external" href="https://github.com/project-gemmi/benchmarking-json">much faster</a>
and parsing CIF and JSON is not that different),
it is not a priority.</p>
</div>
<div class="section" id="directory-walking">
<h2>Directory walking<a class="headerlink" href="#directory-walking" title="Permalink to this headline">¶</a></h2>
<p>Many of the utilities and examples developed for this project
work with archives of CIF files such as wwPDB or COD.
To make it easier to iterate over all CIF files in a directory tree
we provide a class <code class="docutils literal notranslate"><span class="pre">CifWalk</span></code>.</p>
<div class="section" id="id19">
<h3>C++<a class="headerlink" href="#id19" title="Permalink to this headline">¶</a></h3>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="cp">#include</span> <span class="cpf">&lt;gemmi/dirwalk.hpp&gt;</span><span class="cp"></span>

<span class="c1">// ...</span>
<span class="c1">// throws std::runtime_error if top_dir doesn&#39;t exist</span>
<span class="k">for</span> <span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="nl">cif_file</span> <span class="p">:</span> <span class="n">gemmi</span><span class="o">::</span><span class="n">CifWalk</span><span class="p">(</span><span class="n">top_dir</span><span class="p">))</span> <span class="p">{</span>
  <span class="n">cif</span><span class="o">::</span><span class="n">Document</span> <span class="n">doc</span> <span class="o">=</span> <span class="n">cif</span><span class="o">::</span><span class="n">read</span><span class="p">(</span><span class="n">gemmi</span><span class="o">::</span><span class="n">MaybeGzipped</span><span class="p">(</span><span class="n">cif_file</span><span class="p">));</span>
  <span class="c1">// ...</span>
<span class="p">}</span>
</pre></div>
</div>
<p>This header file contains also a more general <code class="docutils literal notranslate"><span class="pre">DirWalk</span></code> class,
and classes specific to macromolecular files (<code class="docutils literal notranslate"><span class="pre">PdbWalk</span></code>, <code class="docutils literal notranslate"><span class="pre">MmCifWalk</span></code>,
<code class="docutils literal notranslate"><span class="pre">CoorFileWalk</span></code>). The file type of each file is guessed from
the file name.</p>
</div>
<div class="section" id="id20">
<h3>Python<a class="headerlink" href="#id20" title="Permalink to this headline">¶</a></h3>
<p>Since Python comes with the os.walk() function for iterating over files
and directories, this functionality is less important here.
Anyway, we provide bindings for CifWalk:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">gemmi</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">gemmi</span><span class="o">.</span><span class="n">CifWalk</span><span class="p">(</span><span class="s1">&#39;../tests/&#39;</span><span class="p">))[:</span><span class="mi">2</span><span class="p">]</span>
<span class="go">[&#39;../tests/1011031.cif&#39;, &#39;../tests/1pfe.cif.gz&#39;]</span>
</pre></div>
</div>
<p>We also have Python bindings for <code class="docutils literal notranslate"><span class="pre">CoorFileWalk</span></code> that picks macromolecular
coordinate files.</p>
<hr class="docutils" />
<p>When the user has no permission to read one of the traversed directories,
the functions above raise an error (std::runtime_error / RuntimeError).</p>
<p>All these directory walking functions are powered by the
<a class="reference external" href="https://github.com/cxong/tinydir">tinydir</a> library
(a single-header library copied into <code class="docutils literal notranslate"><span class="pre">include/gemmi/third_party</span></code>).</p>
</div>
</div>
<div class="section" id="examples">
<span id="cif-examples"></span><h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<p>The examples here use C++11 or Python 2.7/3.x.
Full working code code can be found in the <a class="reference external" href="https://github.com/project-gemmi/gemmi/tree/master/examples">examples</a> directory.</p>
<p>If you have the Python <code class="docutils literal notranslate"><span class="pre">gemmi</span></code> module installed you should also have
the Python examples – try <code class="docutils literal notranslate"><span class="pre">python</span> <span class="pre">-m</span> <span class="pre">gemmi-examples</span></code> if not sure
where they are.</p>
<p>The examples below can be run on one or more PDBx/mmCIF files.
The ones that perform PDB-wide analysis are meant to be run on a
<a class="reference external" href="https://www.wwpdb.org/download/downloads">local copy</a> of the mmCIF
archive (30GB+ gzipped, don’t uncompress!).</p>
<div class="section" id="mmcif-to-xyz">
<h3>mmCIF to XYZ<a class="headerlink" href="#mmcif-to-xyz" title="Permalink to this headline">¶</a></h3>
<p>To start with something simple, let us convert mmCIF
to the <a class="reference external" href="https://en.wikipedia.org/wiki/XYZ_file_format">XYZ format</a>:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="cp">#include</span> <span class="cpf">&lt;iostream&gt;</span><span class="cp"></span>
<span class="cp">#include</span> <span class="cpf">&lt;gemmi/cif.hpp&gt;</span><span class="cp"></span>
<span class="cp">#include</span> <span class="cpf">&lt;gemmi/util.hpp&gt;  // for gemmi::join_str</span><span class="cp"></span>
<span class="c1">//#include &lt;boost/algorithm/string/join.hpp&gt;  // for boost::algorithm::join</span>

<span class="kt">void</span> <span class="nf">convert_to_xyz</span><span class="p">(</span><span class="n">cif</span><span class="o">::</span><span class="n">Document</span> <span class="n">doc</span><span class="p">)</span> <span class="p">{</span>
  <span class="n">cif</span><span class="o">::</span><span class="n">Table</span> <span class="n">atoms</span> <span class="o">=</span> <span class="n">doc</span><span class="p">.</span><span class="n">sole_block</span><span class="p">().</span><span class="n">find</span><span class="p">(</span><span class="s">&quot;atom_site_.&quot;</span><span class="p">,</span>
                          <span class="p">{</span><span class="s">&quot;type_symbol&quot;</span><span class="p">,</span> <span class="s">&quot;Cartn_x&quot;</span><span class="p">,</span> <span class="s">&quot;Cartn_y&quot;</span><span class="p">,</span> <span class="s">&quot;Cartn_z&quot;</span><span class="p">});</span>
  <span class="n">std</span><span class="o">::</span><span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="n">atoms</span><span class="p">.</span><span class="n">length</span><span class="p">()</span> <span class="o">&lt;&lt;</span> <span class="s">&quot;</span><span class="se">\n\n</span><span class="s">&quot;</span><span class="p">;</span>
  <span class="k">for</span> <span class="p">(</span><span class="k">auto</span> <span class="nl">row</span> <span class="p">:</span> <span class="n">atoms</span><span class="p">)</span>
    <span class="n">std</span><span class="o">::</span><span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="n">gemmi</span><span class="o">::</span><span class="n">join_str</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="s">&quot; &quot;</span><span class="p">)</span> <span class="o">&lt;&lt;</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">;</span>
    <span class="c1">//std::cout &lt;&lt; boost::algorithm::join(row, &quot; &quot;) &lt;&lt; &quot;\n&quot;;</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="auth-vs-label">
<span id="auth-label-example"></span><h3>auth vs label<a class="headerlink" href="#auth-vs-label" title="Permalink to this headline">¶</a></h3>
<p>When you look at the list of atoms (<code class="docutils literal notranslate"><span class="pre">_atom_site.*</span></code>) in mmCIF files
some columns seem to be completely redundant. Are they?</p>
<table style="font-size:11px; border-spacing:5px 0; margin-top:60px; width:100%; border-collapse:separate;">
<colgroup>
<col span="3">
<col style="background-color:#fc8">
<col>
<col style="background-color:#ffa">
<col style="background-color:#aff">
<col>
<col style="background-color:#bd8">
<col span="12">
<col style="background-color:#bd8">
<col style="background-color:#ffa">
<col style="background-color:#aff">
<col style="background-color:#fc8">
<col>
</colgroup>
<tr style="font-size:10px;">
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">group_PDB</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">id</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">type_symbol</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">label_atom_id</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">label_alt_id</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">label_comp_id</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">label_asym_id</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">label_entity_id</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">label_seq_id</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">pdbx_PDB_ins_code</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">Cartn_x</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">Cartn_y</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">Cartn_z</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">occupancy</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">B_iso_or_equiv</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">Cartn_x_esd</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">Cartn_y_esd</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">Cartn_z_esd</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">occupancy_esd</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">B_iso_or_equiv_esd</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">pdbx_formal_charge</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">auth_seq_id</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">auth_comp_id</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">auth_asym_id</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">auth_atom_id</th>
<th style="max-width:12px; transform-origin:left; transform:rotate(315deg);">pdbx_PDB_model_num</th>
<tr>
<td>ATOM</td><td>1</td><td>N</td><td>N</td><td>.</td><td>GLY</td><td>A</td>
<td>1</td><td>1</td><td>?</td><td>-9.009</td><td>4.612</td><td>6.102</td>
<td>1.00</td><td>16.77</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>1</td><td>GLY</td><td>A</td><td>N</td><td>1</td>
</tr><tr>
<td>ATOM</td><td>2</td><td>C</td><td>CA</td><td>.</td><td>GLY</td><td>A</td>
<td>1</td><td>1</td><td>?</td><td>-9.052</td><td>4.207</td><td>4.651</td>
<td>1.00</td><td>16.57</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>1</td><td>GLY</td><td>A</td><td>CA</td><td>1</td>
</tr><tr>
<td>ATOM</td><td>3</td><td>C</td><td>C</td><td>.</td><td>GLY</td><td>A</td>
<td>1</td><td>1</td><td>?</td><td>-8.015</td><td>3.140</td><td>4.419</td>
<td>1.00</td><td>16.16</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>1</td><td>GLY</td><td>A</td><td>C</td><td>1</td>
</tr><tr>
<td>ATOM</td><td>4</td><td>O</td><td>O</td><td>.</td><td>GLY</td><td>A</td>
<td>1</td><td>1</td><td>?</td><td>-7.523</td><td>2.521</td><td>5.381</td>
<td>1.00</td><td>16.78</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>1</td><td>GLY</td><td>A</td><td>O</td><td>1</td>
</tr><tr>
<td>ATOM</td><td>5</td><td>N</td><td>N</td><td>.</td><td>ASN</td><td>A</td>
<td>1</td><td>2</td><td>?</td><td>-7.656</td><td>2.923</td><td>3.155</td>
<td>1.00</td><td>15.02</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>2</td><td>ASN</td><td>A</td><td>N</td><td>1</td>
</tr><tr>
<td>ATOM</td><td>6</td><td>C</td><td>CA</td><td>.</td><td>ASN</td><td>A</td>
<td>1</td><td>2</td><td>?</td><td>-6.522</td><td>2.038</td><td>2.831</td>
<td>1.00</td><td>14.10</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>2</td><td>ASN</td><td>A</td><td>CA</td><td>1</td>
</tr><tr>
<td>ATOM</td><td>7</td><td>C</td><td>C</td><td>.</td><td>ASN</td><td>A</td>
<td>1</td><td>2</td><td>?</td><td>-5.241</td><td>2.537</td><td>3.427</td>
<td>1.00</td><td>13.13</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>2</td><td>ASN</td><td>A</td><td>C</td><td>1</td>
</tr><tr>
<td>ATOM</td><td>8</td><td>O</td><td>O</td><td>.</td><td>ASN</td><td>A</td>
<td>1</td><td>2</td><td>?</td><td>-4.978</td><td>3.742</td><td>3.426</td>
<td>1.00</td><td>11.91</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>2</td><td>ASN</td><td>A</td><td>O</td><td>1</td>
</tr><tr>
<td>ATOM</td><td>9</td><td>C</td><td>CB</td><td>.</td><td>ASN</td><td>A</td>
<td>1</td><td>2</td><td>?</td><td>-6.346</td><td>1.881</td><td>1.341</td>
<td>1.00</td><td>15.38</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>2</td><td>ASN</td><td>A</td><td>CB</td><td>1</td>
</tr><tr>
<td>ATOM</td><td>10</td><td>C</td><td>CG</td><td>.</td><td>ASN</td><td>A</td>
<td>1</td><td>2</td><td>?</td><td>-7.584</td><td>1.342</td><td>0.692</td>
<td>1.00</td><td>14.08</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>2</td><td>ASN</td><td>A</td><td>CG</td><td>1</td>
</tr><tr>
<td>ATOM</td><td>11</td><td>O</td><td>OD1</td><td>.</td><td>ASN</td><td>A</td>
<td>1</td><td>2</td><td>?</td><td>-8.025</td><td>0.227</td><td>1.016</td>
<td>1.00</td><td>17.46</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>2</td><td>ASN</td><td>A</td><td>OD1</td><td>1</td>
</tr><tr>
<td>ATOM</td><td>12</td><td>N</td><td>ND2</td><td>.</td><td>ASN</td><td>A</td>
<td>1</td><td>2</td><td>?</td><td>-8.204</td><td>2.155</td><td>-0.169</td>
<td>1.00</td><td>11.72</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>2</td><td>ASN</td><td>A</td><td>ND2</td><td>1</td>
</tr><tr>
<td colspan="25">...</td>
</tr><tr>
<td>ATOM</td><td>58</td><td>O</td><td>OH</td><td>.</td><td>TYR</td><td>A</td>
<td>1</td><td>7</td><td>?</td><td>3.766</td><td>0.589</td><td>10.291</td>
<td>1.00</td><td>14.39</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>7</td><td>TYR</td><td>A</td><td>OH</td><td>1</td>
</tr><tr>
<td>ATOM</td><td>59</td><td>O</td><td>OXT</td><td>.</td><td>TYR</td><td>A</td>
<td>1</td><td>7</td><td>?</td><td>11.358</td><td>2.999</td><td>7.612</td>
<td>1.00</td><td>17.49</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>7</td><td>TYR</td><td>A</td><td>OXT</td><td>1</td>
</tr><tr>
<td>HETATM</td><td>60</td><td>O</td><td>O</td><td>.</td><td>HOH</td><td>B</td>
<td>2</td><td>.</td><td>?</td><td>-6.471</td><td>5.227</td><td>7.124</td>
<td>1.00</td><td>22.62</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>8</td><td>HOH</td><td>A</td><td>O</td><td>1</td>
</tr><tr>
<td>HETATM</td><td>61</td><td>O</td><td>O</td><td>.</td><td>HOH</td><td>B</td>
<td>2</td><td>.</td><td>?</td><td>10.431</td><td>1.858</td><td>3.216</td>
<td>1.00</td><td>19.71</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>9</td><td>HOH</td><td>A</td><td>O</td><td>1</td>
</tr><tr>
<td>HETATM</td><td>62</td><td>O</td><td>O</td><td>.</td><td>HOH</td><td>B</td>
<td>2</td><td>.</td><td>?</td><td>-11.286</td><td>1.756</td><td>-1.468</td>
<td>1.00</td><td>17.08</td><td>?</td><td>?</td><td>?</td><td>?</td><td>?</td>
<td>?</td><td>10</td><td>HOH</td><td>A</td><td>O</td><td>1</td>
</tr><tr>
</tr>
</table><p>It is hard to manually find an example
where <code class="docutils literal notranslate"><span class="pre">_atom_site.auth_atom_id</span></code> differs from <code class="docutils literal notranslate"><span class="pre">_atom_site.label_atom_id</span></code>, or
where <code class="docutils literal notranslate"><span class="pre">_atom_site.auth_comp_id</span></code> differs from <code class="docutils literal notranslate"><span class="pre">_atom_site.label_comp_id</span></code>.
So here is a small C++ program:</p>
<div class="highlight-cpp notranslate"><div class="highlight"><pre><span></span><span class="c1">// Compare pairs of columns from the _atom_site table.</span>
<span class="c1">// Compiled with: g++-6 -O2 -Iinclude auth_label.cc -lstdc++fs -lz</span>
<span class="cp">#include</span> <span class="cpf">&lt;gemmi/gz.hpp&gt;</span><span class="cp"></span>
<span class="cp">#include</span> <span class="cpf">&lt;gemmi/cif.hpp&gt;</span><span class="cp"></span>
<span class="cp">#include</span> <span class="cpf">&lt;iostream&gt;</span><span class="cp"></span>
<span class="cp">#include</span> <span class="cpf">&lt;experimental/filesystem&gt;  // just &lt;filesystem&gt; in C++17</span><span class="cp"></span>

<span class="k">namespace</span> <span class="n">fs</span> <span class="o">=</span> <span class="n">std</span><span class="o">::</span><span class="n">experimental</span><span class="o">::</span><span class="n">filesystem</span><span class="p">;</span>
<span class="k">namespace</span> <span class="n">cif</span> <span class="o">=</span> <span class="n">gemmi</span><span class="o">::</span><span class="n">cif</span><span class="p">;</span>

<span class="kt">void</span> <span class="nf">print_differences</span><span class="p">(</span><span class="n">cif</span><span class="o">::</span><span class="n">Block</span><span class="o">&amp;</span> <span class="n">block</span><span class="p">,</span> <span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">name</span><span class="p">)</span> <span class="p">{</span>
  <span class="k">for</span> <span class="p">(</span><span class="k">auto</span> <span class="nl">row</span> <span class="p">:</span> <span class="n">block</span><span class="p">.</span><span class="n">find</span><span class="p">(</span><span class="s">&quot;_atom_site.&quot;</span><span class="p">,</span> <span class="p">{</span><span class="s">&quot;auth_&quot;</span> <span class="o">+</span> <span class="n">name</span><span class="p">,</span> <span class="s">&quot;label_&quot;</span> <span class="o">+</span> <span class="n">name</span><span class="p">}))</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
      <span class="n">std</span><span class="o">::</span><span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="n">block</span><span class="p">.</span><span class="n">name</span> <span class="o">&lt;&lt;</span> <span class="s">&quot;: &quot;</span> <span class="o">&lt;&lt;</span> <span class="n">name</span> <span class="o">&lt;&lt;</span> <span class="s">&quot;  &quot;</span>
                <span class="o">&lt;&lt;</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;&lt;</span> <span class="s">&quot; -&gt; &quot;</span> <span class="o">&lt;&lt;</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;&lt;</span> <span class="n">std</span><span class="o">::</span><span class="n">endl</span><span class="p">;</span>
<span class="p">}</span>

<span class="kt">bool</span> <span class="nf">ends_with</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">str</span><span class="p">,</span> <span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">suffix</span><span class="p">)</span> <span class="p">{</span>
  <span class="kt">size_t</span> <span class="n">sl</span> <span class="o">=</span> <span class="n">suffix</span><span class="p">.</span><span class="n">length</span><span class="p">();</span>
  <span class="k">return</span> <span class="n">str</span><span class="p">.</span><span class="n">length</span><span class="p">()</span> <span class="o">&gt;=</span> <span class="n">sl</span> <span class="o">&amp;&amp;</span> <span class="n">str</span><span class="p">.</span><span class="n">compare</span><span class="p">(</span><span class="n">str</span><span class="p">.</span><span class="n">length</span><span class="p">()</span> <span class="o">-</span> <span class="n">sl</span><span class="p">,</span> <span class="n">sl</span><span class="p">,</span> <span class="n">suffix</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">;</span>
<span class="p">}</span>

<span class="kt">int</span> <span class="nf">main</span><span class="p">(</span><span class="kt">int</span> <span class="n">argc</span><span class="p">,</span> <span class="kt">char</span><span class="o">*</span> <span class="n">argv</span><span class="p">[])</span> <span class="p">{</span>
  <span class="k">if</span> <span class="p">(</span><span class="n">argc</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">)</span>
    <span class="k">return</span> <span class="mi">1</span><span class="p">;</span>
  <span class="k">for</span> <span class="p">(</span><span class="k">auto</span><span class="o">&amp;</span> <span class="nl">p</span> <span class="p">:</span> <span class="n">fs</span><span class="o">::</span><span class="n">recursive_directory_iterator</span><span class="p">(</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span> <span class="p">{</span>
    <span class="n">std</span><span class="o">::</span><span class="n">string</span> <span class="n">path</span> <span class="o">=</span> <span class="n">p</span><span class="p">.</span><span class="n">path</span><span class="p">().</span><span class="n">u8string</span><span class="p">();</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">ends_with</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s">&quot;.cif&quot;</span><span class="p">)</span> <span class="o">||</span> <span class="n">ends_with</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s">&quot;.cif.gz&quot;</span><span class="p">))</span> <span class="p">{</span>
      <span class="n">cif</span><span class="o">::</span><span class="n">Document</span> <span class="n">doc</span> <span class="o">=</span> <span class="n">cif</span><span class="o">::</span><span class="n">read</span><span class="p">(</span><span class="n">gemmi</span><span class="o">::</span><span class="n">MaybeGzipped</span><span class="p">(</span><span class="n">path</span><span class="p">));</span>
      <span class="c1">// What author&#39;s atom names were changed?</span>
      <span class="n">print_differences</span><span class="p">(</span><span class="n">doc</span><span class="p">.</span><span class="n">sole_block</span><span class="p">(),</span> <span class="s">&quot;atom_id&quot;</span><span class="p">);</span>
      <span class="c1">// What author&#39;s residue names were changed?</span>
      <span class="n">print_differences</span><span class="p">(</span><span class="n">doc</span><span class="p">.</span><span class="n">sole_block</span><span class="p">(),</span> <span class="s">&quot;comp_id&quot;</span><span class="p">);</span>
    <span class="p">}</span>
  <span class="p">}</span>
  <span class="k">return</span> <span class="mi">0</span><span class="p">;</span>
<span class="p">}</span>
</pre></div>
</div>
<p>We compile it, run it, and come back after an hour:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>$ g++-6 -O2 -Iinclude examples/auth_label.cpp -lstdc++fs -lz
$ ./a.out pdb_copy/mmCIF
3D3W: atom_id  O1 -&gt; OD
1TNI: atom_id  HN2 -&gt; HN3
1S01: comp_id  CA -&gt; UNL
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
2KI7: atom_id  H -&gt; H1
4E1U: atom_id  H101 -&gt; H103
4WH8: atom_id  H3 -&gt; H391
4WH8: atom_id  H4 -&gt; H401
3V2I: atom_id  UNK -&gt; UNL
1AGG: atom_id  H3 -&gt; H
1AGG: atom_id  H3 -&gt; H
1AGG: atom_id  H3 -&gt; H
</pre></div>
</div>
<p>So, as of April 2017, only a single author’s residue name was changed,
and atom names were changed in 7 PDB entries.</p>
</div>
<div class="section" id="amino-acid-frequency">
<h3>Amino acid frequency<a class="headerlink" href="#amino-acid-frequency" title="Permalink to this headline">¶</a></h3>
<p>After seeing a <a class="reference external" href="https://doi.org/10.1093/nar/gkw978">paper</a>
in which amino acid frequency was averaged over 5000+ proteomes
(Ala 8.76%, Cys 1.38%, Asp 5.49%, etc)
we may want to compare it with the frequency in the PDB database.
So we write a little script</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/usr/bin/env python</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">print_function</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">Counter</span>
<span class="kn">from</span> <span class="nn">gemmi</span> <span class="kn">import</span> <span class="n">cif</span>

<span class="c1"># To keep this example small we moved handling of command-line args to util.py.</span>
<span class="c1"># When a directory is given as an argument get_file_paths_from_args()</span>
<span class="c1"># yields all the cif(.gz) paths under this directory.</span>
<span class="kn">from</span> <span class="nn">util</span> <span class="kn">import</span> <span class="n">get_file_paths_from_args</span>

<span class="c1"># Check amino-acid frequency in the PDB database</span>
<span class="n">totals</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">()</span>
<span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">get_file_paths_from_args</span><span class="p">():</span>
    <span class="c1"># read file (uncompressing on the fly) and get the only block</span>
    <span class="n">block</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="o">.</span><span class="n">sole_block</span><span class="p">()</span>
    <span class="c1"># find table with the sequence</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;_entity_poly_seq.&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;entity_id&#39;</span><span class="p">,</span> <span class="s1">&#39;mon_id&#39;</span><span class="p">])</span>
    <span class="c1"># convert table with chain types (protein/DNA/RNA) to dict</span>
    <span class="n">entity_types</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;_entity_poly.&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;entity_id&#39;</span><span class="p">,</span> <span class="s1">&#39;type&#39;</span><span class="p">]))</span>
    <span class="c1"># and count these monomers that correspond to a protein chain</span>
    <span class="n">aa_counter</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">seq</span>
                         <span class="k">if</span> <span class="s1">&#39;polypeptide&#39;</span> <span class="ow">in</span> <span class="n">entity_types</span><span class="p">[</span><span class="n">row</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="mi">0</span><span class="p">)])</span>
    <span class="n">totals</span> <span class="o">+=</span> <span class="n">aa_counter</span>
    <span class="c1"># print residue counts for each file</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="o">*</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1">:</span><span class="si">%d</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">aa_counter</span><span class="o">.</span><span class="n">most_common</span><span class="p">()))</span>
<span class="c1"># finally, print the total counts as percentages</span>
<span class="n">f</span> <span class="o">=</span> <span class="mf">100.0</span> <span class="o">/</span> <span class="nb">sum</span><span class="p">(</span><span class="n">totals</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;TOTAL&#39;</span><span class="p">,</span> <span class="o">*</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1">:</span><span class="si">%.2f%%</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">c</span><span class="o">*</span><span class="n">f</span><span class="p">)</span> <span class="k">for</span> <span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">c</span><span class="p">)</span> <span class="ow">in</span> <span class="n">totals</span><span class="o">.</span><span class="n">most_common</span><span class="p">(</span><span class="mi">20</span><span class="p">)))</span>
</pre></div>
</div>
<p>We can run this script on our copy of the PDB database
(mmCIF format) and get such an output:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>200L ALA:17 LEU:15 LYS:13 ARG:13 THR:12 ASN:12 GLY:11 ILE:10 ASP:10 VAL:9 GLU:8 SER:6 TYR:6 GLN:5 PHE:5 MET:5 PRO:3 TRP:3 HIS:1
...
4ZZZ LEU:40 LYS:33 SER:30 ASP:25 GLY:25 ILE:25 VAL:23 ALA:19 PRO:18 GLU:18 ASN:17 THR:16 TYR:16 GLN:14 ARG:11 PHE:10 HIS:9 MET:9 CYS:2 TRP:2
TOTAL LEU:8.90% ALA:7.80% GLY:7.34% VAL:6.95% GLU:6.55% SER:6.29% LYS:6.18% ASP:5.55% THR:5.54% ILE:5.49% ARG:5.35% PRO:4.66% ASN:4.16% PHE:3.88% GLN:3.77% TYR:3.45% HIS:2.63% MET:2.14% CYS:1.38% TRP:1.35%
</pre></div>
</div>
<p>On my laptop it takes about an hour, using a single core.
Most of this hour is spent on tokenizing the CIF files and copying
the content into a DOM structure, what could be largely avoided given
that we use only sequences not atoms.
But it is not worth to optimize one-off scripts.
The same goes for using multiple processor cores.</p>
</div>
<div class="section" id="custom-pdb-search">
<h3>Custom PDB search<a class="headerlink" href="#custom-pdb-search" title="Permalink to this headline">¶</a></h3>
<p>We may need to go through a local copy of the PDB archive to find entries
according to criteria that cannot be queried in RCSB/PDBe/PDBj web interfaces.
For example, if for some reason we need PDB entries with large number
of anisotropic B-factors, we may write a quick script:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/usr/bin/env python</span>
<span class="s2">&quot;Find PDB entries with more than 50,000 anisotropic B-factors.&quot;</span>

<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">print_function</span>
<span class="kn">from</span> <span class="nn">gemmi</span> <span class="kn">import</span> <span class="n">cif</span>
<span class="kn">from</span> <span class="nn">util</span> <span class="kn">import</span> <span class="n">get_file_paths_from_args</span>

<span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">get_file_paths_from_args</span><span class="p">():</span>
    <span class="n">block</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="o">.</span><span class="n">sole_block</span><span class="p">()</span>
    <span class="n">anis</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find_values</span><span class="p">(</span><span class="s2">&quot;_atom_site_anisotrop.id&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">anis</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">50000</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">anis</span><span class="p">))</span>
</pre></div>
</div>
<p>and then run it for an hour or so.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>$ ./examples/simple_search.py pdb_copy/mmCIF
5AEW 60155
5AFI 98325
3AIB 52592
...
</pre></div>
</div>
</div>
<div class="section" id="search-pdb-by-elements">
<h3>Search PDB by elements<a class="headerlink" href="#search-pdb-by-elements" title="Permalink to this headline">¶</a></h3>
<a class="reference external image-reference" href="https://project-gemmi.github.io/periodic-table/"><img alt="_images/periodic-table-thumb.png" class="align-right" src="_images/periodic-table-thumb.png" style="width: 171.0px; height: 171.0px;" /></a>
<p>Let say we want to be able to search the PDB by specifying a set of elements
present in the model. First we write down elements present in each
PDB entry:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">block</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="o">.</span><span class="n">sole_block</span><span class="p">()</span>
<span class="n">elems</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">find_loop</span><span class="p">(</span><span class="s2">&quot;_atom_site.type_symbol&quot;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">name</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">elems</span><span class="p">))</span>
</pre></div>
</div>
<p>This example ended up overdone a bit and it was put into a
<a class="reference external" href="https://github.com/project-gemmi/periodic-table">separate repository</a>.</p>
<p>Here is a demo: <a class="reference external" href="https://project-gemmi.github.io/periodic-table/">https://project-gemmi.github.io/periodic-table/</a></p>
</div>
<div class="section" id="solvent-content-vs-resolution">
<h3>Solvent content vs resolution<a class="headerlink" href="#solvent-content-vs-resolution" title="Permalink to this headline">¶</a></h3>
<p>Let say that we would like to generate a plot of solvent content
as a function of <em>d</em><sub>min</sub>, similar to the plots
by C. X. Weichenberger and B. Rupp
in <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/24914969">Acta Cryst D</a>
and <a class="reference external" href="http://www.phenix-online.org/newsletter/CCN_2015_01.pdf#page=14">CNN</a>),
but less smoothed and with duplicate entries included.</p>
<p>In macromolecular crystallography solvent content is conventionally
estimated as:</p>
<blockquote>
<div><p><em>V</em><sub>S</sub> = 1 – 1.230 / <em>V</em><sub>M</sub></p>
</div></blockquote>
<p>where <em>V</em><sub>M</sub> is Matthews coefficient defined as <em>V</em><sub>M</sub>=<em>V</em>/<em>m</em>
(volume of the asymmetric unit over the molecular weight of all
molecules in this volume).</p>
<p>Weichenberger and Rupp calculated <em>V</em><sub>S</sub> themselves, but we will simply
use the values present in mmCIF files.
So first we extract <em>V</em><sub>M</sub>, <em>V</em><sub>S</sub> and <em>d</em><sub>min</sub>,
as well as number of DNA/RNA chains (protein-only entries will have 0 here),
deposition date and group ID (which will be explained later).</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">gather_data</span><span class="p">():</span>
    <span class="s2">&quot;read mmCIF files and write down a few numbers (one file -&gt; one line)&quot;</span>
    <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">,</span> <span class="n">dialect</span><span class="o">=</span><span class="s1">&#39;excel-tab&#39;</span><span class="p">)</span>
    <span class="n">writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="s1">&#39;code&#39;</span><span class="p">,</span> <span class="s1">&#39;na_chains&#39;</span><span class="p">,</span> <span class="s1">&#39;vs&#39;</span><span class="p">,</span> <span class="s1">&#39;vm&#39;</span><span class="p">,</span> <span class="s1">&#39;d_min&#39;</span><span class="p">,</span> <span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="s1">&#39;group&#39;</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">util</span><span class="o">.</span><span class="n">get_file_paths_from_args</span><span class="p">():</span>
        <span class="n">block</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="o">.</span><span class="n">sole_block</span><span class="p">()</span>
        <span class="n">code</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">as_string</span><span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">find_value</span><span class="p">(</span><span class="s1">&#39;_entry.id&#39;</span><span class="p">))</span>
        <span class="n">na</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="s1">&#39;nucleotide&#39;</span> <span class="ow">in</span> <span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                 <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">block</span><span class="o">.</span><span class="n">find_values</span><span class="p">(</span><span class="s1">&#39;_entity_poly.type&#39;</span><span class="p">))</span>
        <span class="n">vs</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find_value</span><span class="p">(</span><span class="s1">&#39;_exptl_crystal.density_percent_sol&#39;</span><span class="p">)</span>
        <span class="n">vm</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find_value</span><span class="p">(</span><span class="s1">&#39;_exptl_crystal.density_Matthews&#39;</span><span class="p">)</span>
        <span class="n">d_min</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find_value</span><span class="p">(</span><span class="s1">&#39;_refine.ls_d_res_high&#39;</span><span class="p">)</span>
        <span class="n">dep_date_tag</span> <span class="o">=</span> <span class="s1">&#39;_pdbx_database_status.recvd_initial_deposition_date&#39;</span>
        <span class="n">dep_date</span> <span class="o">=</span> <span class="n">parse_date</span><span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">find_values</span><span class="p">(</span><span class="n">dep_date_tag</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="mi">0</span><span class="p">))</span>
        <span class="n">group</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find_value</span><span class="p">(</span><span class="s1">&#39;_pdbx_deposit_group.group_id&#39;</span><span class="p">)</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="n">code</span><span class="p">,</span> <span class="n">na</span><span class="p">,</span> <span class="n">vs</span><span class="p">,</span> <span class="n">vm</span><span class="p">,</span> <span class="n">d_min</span><span class="p">,</span> <span class="n">dep_date</span><span class="p">,</span> <span class="n">group</span><span class="p">])</span>
</pre></div>
</div>
<p>After a few checks (see <code class="file docutils literal notranslate"><span class="pre">examples/matthews.py</span></code>) we may decide to use
only those PDB entries in which <em>V</em><sub>M</sub> and <em>V</em><sub>S</sub> are consistent.
To make things more interesting we plot separately depositions from before
and after 1/1/2015.</p>
<div class="figure align-center" id="id22">
<a class="reference internal image-reference" href="_images/solvent-content-old.png"><img alt="A plot of solvent content vs resolution. PDB entries up to 2014." src="_images/solvent-content-old.png" style="width: 636.0px; height: 550.0px;" /></a>
<p class="caption"><span class="caption-text">About 90,000 PDB entries (up to 2014)
plotted as intentionally undersmoothed kernel density estimate.
The code used to produce this plot is in <code class="file docutils literal notranslate"><span class="pre">examples/matthews.py</span></code>.</span><a class="headerlink" href="#id22" title="Permalink to this image">¶</a></p>
</div>
<p>Ripples in the right subplot show that in many entries <em>V</em><sub>S</sub>
is reported as integer, so we should calculate it ourselves
(like Weichenberger &amp; Rupp) for better precision.
Ripples in the top plot show that we should use a less arbitrary metric
of resolution than <em>d</em><sub>min</sub> (but it’s not so easy).
Or we could just smooth them out by changing parameters of this plot.</p>
<div class="figure align-center" id="id23">
<a class="reference internal image-reference" href="_images/solvent-content-recent.png"><img alt="A plot of solvent content vs resolution. PDB entries since 2015." src="_images/solvent-content-recent.png" style="width: 636.0px; height: 550.0px;" /></a>
<p class="caption"><span class="caption-text">About 17,000 PDB entries (2015 - April 2017)
plotted as intentionally undersmoothed kernel density estimate.
The code used to produce this plot is in <code class="file docutils literal notranslate"><span class="pre">examples/matthews.py</span></code>.</span><a class="headerlink" href="#id23" title="Permalink to this image">¶</a></p>
</div>
<p>On the left side of the yellow egg you can see dark stripes
caused by <em>group depositions</em>, which were introduced by PDB in 2016.
They came from two European high-throughput beamlines and
serve as an illustration of how automated software can analyze hundreds
of similar samples (fragment screening) and submit them quickly to the PDB.</p>
<p>We can easily filter out group depositions – either using the group IDs
extracted from mmCIF files (we do this for <a class="reference external" href="https://project-gemmi.github.io/pdb-stats/">pdb-stats</a>)
or, like W&amp;B, by excluding redundant entries based on the unit cell and <em>V</em><sub>M</sub>.</p>
<p>Not all the dark spots are group depositions.
For example, the one at <em>V</em><sub>S</sub>≈66.5%, <em>d</em><sub>min</sub> 2.5-3A is proteasome 20S
studied over years by Huber <em>et al</em>, with dozens PDB submissions.</p>
</div>
<div class="section" id="weights">
<h3>Weights<a class="headerlink" href="#weights" title="Permalink to this headline">¶</a></h3>
<p>Let say we would like to verify consistency of molecular weights.
First, let us look at chem_comp tables:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>loop_
_chem_comp.id
_chem_comp.type
_chem_comp.mon_nstd_flag
_chem_comp.name
_chem_comp.pdbx_synonyms
_chem_comp.formula
_chem_comp.formula_weight
ALA &#39;L-peptide linking&#39; y ALANINE         ?                 &#39;C3 H7 N O2&#39;     89.093
ARG &#39;L-peptide linking&#39; y ARGININE        ?                 &#39;C6 H15 N4 O2 1&#39; 175.209
ASN &#39;L-peptide linking&#39; y ASPARAGINE      ?                 &#39;C4 H8 N2 O3&#39;    132.118
ASP &#39;L-peptide linking&#39; y &#39;ASPARTIC ACID&#39; ?                 &#39;C4 H7 N O4&#39;     133.103
CYS &#39;L-peptide linking&#39; y CYSTEINE        ?                 &#39;C3 H7 N O2 S&#39;   121.158
EDO non-polymer         . 1,2-ETHANEDIOL  &#39;ETHYLENE GLYCOL&#39; &#39;C2 H6 O2&#39;       62.068
...
</pre></div>
</div>
<p>We expect that by using molecular weights of elements and a simple arithmetic
we can recalculate <code class="docutils literal notranslate"><span class="pre">_chem_comp.formula_weight</span></code> from <code class="docutils literal notranslate"><span class="pre">_chem_comp.formula</span></code>.
The full code is in <code class="file docutils literal notranslate"><span class="pre">examples/weights.py`</span></code>.
It includes a function that converts <code class="docutils literal notranslate"><span class="pre">'C2</span> <span class="pre">H6</span> <span class="pre">O2'</span></code> to <code class="docutils literal notranslate"><span class="pre">{C:2,</span> <span class="pre">H:6,</span> <span class="pre">O:2}</span></code>.
Here we only show the few lines of code that sum the element weights
and compare the result:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">check_chem_comp_formula_weight</span><span class="p">(</span><span class="n">block</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">cc</span> <span class="ow">in</span> <span class="n">block</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;_chem_comp.&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;formula&#39;</span><span class="p">,</span> <span class="s1">&#39;formula_weight&#39;</span><span class="p">]):</span>
        <span class="k">if</span> <span class="n">cc</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;UNX&#39;</span><span class="p">,</span> <span class="s1">&#39;UNL&#39;</span><span class="p">):</span>  <span class="c1"># unknown residue or ligand</span>
            <span class="k">continue</span>
        <span class="n">fdict</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">formula_to_dict</span><span class="p">(</span><span class="n">cc</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">title</span><span class="p">())</span>
        <span class="n">calc_weight</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">n</span> <span class="o">*</span> <span class="n">Element</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="o">.</span><span class="n">weight</span> <span class="k">for</span> <span class="p">(</span><span class="n">e</span><span class="p">,</span> <span class="n">n</span><span class="p">)</span> <span class="ow">in</span> <span class="n">fdict</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
        <span class="n">diff</span> <span class="o">=</span> <span class="n">calc_weight</span> <span class="o">-</span> <span class="n">cif</span><span class="o">.</span><span class="n">as_number</span><span class="p">(</span><span class="n">cc</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mf">0.1</span><span class="p">):</span>  <span class="c1"># also true if diff is NaN</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1">  </span><span class="si">%-16s</span><span class="s1"> </span><span class="si">% 9.3f</span><span class="s1"> - </span><span class="si">%9s</span><span class="s1"> = </span><span class="si">%+.3f</span><span class="s1">&#39;</span> <span class="o">%</span>
                  <span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">cc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">cc</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">calc_weight</span><span class="p">,</span> <span class="n">cc</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">diff</span><span class="p">))</span>
</pre></div>
</div>
<p>This script prints differences above 0.1 u:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>3A1R DOD  D2 O                20.028 -    18.015 = +2.013
5AI2 D8U  D 1                  2.014 -         ? = +nan
5AI2 DOD  D2 O                20.028 -    18.015 = +2.013
4AR3 D3O  O 1                 15.999 -    22.042 = -6.043
4AR4 D3O  O 1                 15.999 -    22.042 = -6.043
4AR5 DOD  D2 O                20.028 -    18.015 = +2.013
4AR6 DOD  D2 O                20.028 -    18.015 = +2.013
4B1A K3G  MO12 O40 P 1      1822.350 -  1822.230 = +0.120
4BVO E43  H2 O40 W12 6      2848.072 -  2848.192 = -0.120
...
5FHW HFW  Hf O61 P2 W17     4341.681 -  4343.697 = -2.016
...
</pre></div>
</div>
<p>The differences are few and minor.
We see a few PDB entries with the weight of D<sub>2</sub>O
set to the weight H<sub>2</sub>O. The second line shows missing weight.
The differences +0.12 and -0.12 next to Mo12 and W12 probably come from
the 0.01u difference in the input masses of the elements.
In two entries D3O is missing D in the formula, and -2.016 in HFW
suggests two missing hydrogens.</p>
<p>Now let us try to re-calculate <code class="docutils literal notranslate"><span class="pre">_entity.formula_weight</span></code> from the chem_comp
weights and the sequence.
The PDB software calculates it as a sum of components in the chain,
minus the weight of N-1 waters.
In case of nucleic acids also PO<sub>2</sub> is subtracted
(why not PO<sub>3</sub>? – to be checked).
And in case of microheterogeneity only the main conformer is taken into
account. As the PDB software uses single precision for these computations,
we ignore differences below 0.003%, which we checked to be enough to
account for numerical errors.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">check_entity_formula_weight</span><span class="p">(</span><span class="n">block</span><span class="p">):</span>
    <span class="c1"># read chem_comp weights from the file, e.g. THR: 119.120</span>
    <span class="n">cc_weights</span> <span class="o">=</span> <span class="p">{</span><span class="n">cc</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="mi">0</span><span class="p">):</span> <span class="n">cif</span><span class="o">.</span><span class="n">as_number</span><span class="p">(</span><span class="n">cc</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">cc</span> <span class="ow">in</span>
                  <span class="n">block</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;_chem_comp.&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;formula_weight&#39;</span><span class="p">])}</span>

    <span class="c1"># read polymer types from _entity_poly.type</span>
    <span class="n">poly_types</span> <span class="o">=</span> <span class="p">{</span><span class="n">ep</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="mi">0</span><span class="p">):</span> <span class="n">ep</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">ep</span> <span class="ow">in</span>
                  <span class="n">block</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;_entity_poly.&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;entity_id&#39;</span><span class="p">,</span> <span class="s1">&#39;type&#39;</span><span class="p">])}</span>

    <span class="c1"># read and sort sequences from _entity_poly_seq</span>
    <span class="n">entity_seq</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">block</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;_entity_poly_seq.&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;entity_id&#39;</span><span class="p">,</span> <span class="s1">&#39;num&#39;</span><span class="p">,</span> <span class="s1">&#39;mon_id&#39;</span><span class="p">]):</span>
        <span class="n">entity_seq</span><span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="mi">0</span><span class="p">)]</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">cif</span><span class="o">.</span><span class="n">as_int</span><span class="p">(</span><span class="n">m</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="n">cif</span><span class="o">.</span><span class="n">as_string</span><span class="p">(</span><span class="n">m</span><span class="p">[</span><span class="mi">2</span><span class="p">])))</span>
    <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">entity_seq</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="n">seq</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">p</span><span class="p">:</span> <span class="n">p</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="c1"># read _entity.formula_weight values</span>
    <span class="n">f_weights</span> <span class="o">=</span> <span class="n">block</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;_entity.&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;formula_weight&#39;</span><span class="p">])</span>
    <span class="n">entity_weights</span> <span class="o">=</span> <span class="p">{</span><span class="n">ent</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="mi">0</span><span class="p">):</span> <span class="n">cif</span><span class="o">.</span><span class="n">as_number</span><span class="p">(</span><span class="n">ent</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">ent</span> <span class="ow">in</span> <span class="n">f_weights</span><span class="p">}</span>

    <span class="c1"># calculate weight from sequences and compare with _entity.formula_weight</span>
    <span class="k">for</span> <span class="n">ent</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">entity_seq</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="c1"># in case of microheterogeneity take the first conformer</span>
        <span class="n">main_seq</span> <span class="o">=</span> <span class="p">[</span><span class="nb">next</span><span class="p">(</span><span class="n">g</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">p</span><span class="p">:</span> <span class="n">p</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span>
        <span class="n">weight</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">cc_weights</span><span class="p">[</span><span class="n">mon_id</span><span class="p">]</span> <span class="k">for</span> <span class="p">(</span><span class="n">num</span><span class="p">,</span> <span class="n">mon_id</span><span class="p">)</span> <span class="ow">in</span> <span class="n">main_seq</span><span class="p">)</span>
        <span class="n">weight</span> <span class="o">-=</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">main_seq</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">H2O_MASS</span>
        <span class="k">if</span> <span class="s1">&#39;ribonucleotide&#39;</span> <span class="ow">in</span> <span class="n">poly_types</span><span class="p">[</span><span class="n">ent</span><span class="p">]:</span>  <span class="c1"># DNA or RNA</span>
            <span class="n">weight</span> <span class="o">-=</span> <span class="n">PO2_MASS</span>
        <span class="n">diff</span> <span class="o">=</span> <span class="n">weight</span> <span class="o">-</span> <span class="n">entity_weights</span><span class="p">[</span><span class="n">ent</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span> <span class="o">&gt;</span> <span class="nb">max</span><span class="p">(</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">3e-5</span> <span class="o">*</span> <span class="n">weight</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%4s</span><span class="s1"> entity_id: </span><span class="si">%2s</span><span class="s1"> </span><span class="si">%10.2f</span><span class="s1"> - </span><span class="si">%10.2f</span><span class="s1"> = </span><span class="si">%+8.3f</span><span class="s1">&#39;</span> <span class="o">%</span>
                  <span class="p">(</span><span class="n">block</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">ent</span><span class="p">,</span> <span class="n">weight</span><span class="p">,</span> <span class="n">entity_weights</span><span class="p">[</span><span class="n">ent</span><span class="p">],</span> <span class="n">diff</span><span class="p">))</span>
</pre></div>
</div>
<p>Running this script on a local copy of the PDB database prints 26 lines,
and the difference is always (except for 4PMN) the mass of PO<sub>2</sub>
showing that we have not fully reproduced the rule when to subtract this group.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>...
4D67 entity_id: 44 1625855.77 - 1625917.62 =  -61.855
1HZS entity_id:  1    1733.88 -    1796.85 =  -62.973
2K4G entity_id:  1    2158.21 -    2221.18 =  -62.974
3MBS entity_id:  1    2261.35 -    2324.33 =  -62.974
1NR8 entity_id:  2    2883.07 -    2946.05 =  -62.974
3OK2 entity_id:  1    3417.14 -    3354.17 =  +62.968
...
</pre></div>
</div>
</div>
<div class="section" id="disulfide-bonds">
<h3>Disulfide bonds<a class="headerlink" href="#disulfide-bonds" title="Permalink to this headline">¶</a></h3>
<p>If we were curious what residues take part in disulfide bonds we could
write a little script that inspects annotation in the _struct_conn category.
But to show something else, here we will use <code class="docutils literal notranslate"><span class="pre">gemmi</span> <span class="pre">grep</span></code>, a little
utility that is documented in a <a class="reference internal" href="utils.html#grep"><span class="std std-ref">separate section</span></a>.</p>
<p>First we try how to extract interesting data from a single entry:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> gemmi grep --delimiter<span class="o">=</span><span class="s1">&#39; &#39;</span> _struct_conn.conn_type_id <span class="se">\</span>
&gt; -a _struct_conn.ptnr1_label_comp_id -a _struct_conn.ptnr1_label_atom_id <span class="se">\</span>
&gt; -a _struct_conn.ptnr2_label_comp_id -a _struct_conn.ptnr2_label_atom_id <span class="se">\</span>
&gt; 5CBL
<span class="go">5CBL disulf CYS SG BME S2</span>
<span class="go">5CBL covale ILE CD1 BME C2</span>
</pre></div>
</div>
<p>Then we pipe the output through Unix shell utilities.
<code class="docutils literal notranslate"><span class="pre">|</span> <span class="pre">grep</span> <span class="pre">disulf</span></code> limits the output to the disulfide bonds.
<code class="docutils literal notranslate"><span class="pre">|</span> <span class="pre">awk</span> <span class="pre">'{</span> <span class="pre">print</span> <span class="pre">$3,</span> <span class="pre">$4</span> <span class="pre">&quot;\n&quot;</span> <span class="pre">$5,</span> <span class="pre">$6</span> <span class="pre">}'</span></code> changes each line into two;
the first output line above becomes:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>CYS SG
BME S2
</pre></div>
</div>
<p>Then we run it on the whole PDB archive, sort, count and print the
statistics. The complete command is:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> gemmi grep --delimiter<span class="o">=</span><span class="s1">&#39; &#39;</span> _struct_conn.conn_type_id <span class="se">\</span>
&gt; -a _struct_conn.ptnr1_label_comp_id -a _struct_conn.ptnr1_label_atom_id <span class="se">\</span>
&gt; -a _struct_conn.ptnr2_label_comp_id -a _struct_conn.ptnr2_label_atom_id <span class="se">\</span>
&gt; /hdd/mmCIF <span class="se">\</span>
&gt; <span class="p">|</span> grep disulf <span class="p">|</span> awk <span class="s1">&#39;{ print $3, $4 &quot;\n&quot; $5, $6 }&#39;</span> <span class="se">\</span>
&gt; <span class="p">|</span> sort <span class="p">|</span> uniq -c <span class="p">|</span> sort -nr
<span class="go">   367980 CYS SG</span>
<span class="go">    274 DCY SG</span>
<span class="go">     28 BME S2</span>
<span class="go">     23 SO4 S</span>
<span class="go">     19 CY3 SG</span>
<span class="go">     14 MET SD</span>
<span class="go">     13 MTN S1</span>
<span class="go">     12 V1A S3</span>
<span class="go">     10 MPT SG</span>
<span class="go">      8 NCY SG</span>
<span class="go">      7 LE1 SG</span>
<span class="go">      7 CSX SG</span>
<span class="go">      6 CSO SG</span>
<span class="go">      5 GSH SG2</span>
<span class="go">      5 DTT S1</span>
<span class="go">      4 SC2 SG</span>
<span class="go">      4 MRG S24</span>
<span class="go">      3 H2S S</span>
<span class="go">      3 DTT S4</span>
<span class="go">      3 1WD S7</span>
<span class="go">      2 P8S S1</span>
<span class="go">      2 MTE S2&#39;</span>
<span class="go">      2 MTE S1&#39;</span>
<span class="go">      2 EPE S</span>
<span class="go">      2 DHL SG</span>
<span class="go">      2 DCD S1</span>
<span class="go">      2 CSD SG</span>
<span class="go">      2 COM S1</span>
<span class="go">      2 6LN S2</span>
<span class="go">      2 6LN S1</span>
<span class="go">      2 4K3 S4</span>
<span class="go">      2 3C7 S01</span>
<span class="go">      2 2ON S2</span>
<span class="go">      1 SCN S</span>
<span class="go">      1 RXR SD</span>
<span class="go">      1 RXR S10</span>
<span class="go">      1 MEE S</span>
<span class="go">      1 G47 SG</span>
<span class="go">      1 COA S1P</span>
<span class="go">      1 6ML S1</span>
<span class="go">      1 5O8 SBH</span>
</pre></div>
</div>
<p>And what other bond types are annotated in <code class="docutils literal notranslate"><span class="pre">_struct_conn</span></code>?</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$</span> gemmi grep -O -b _struct_conn.conn_type_id /hdd/mmCIF/ <span class="p">|</span> sort <span class="p">|</span> uniq -c
<span class="go"> 501851 covale</span>
<span class="go"> 184231 disulf</span>
<span class="go">4035115 hydrog</span>
<span class="go">1394732 metalc</span>
</pre></div>
</div>
</div>
<div class="section" id="mmjson-like-data">
<h3>mmJSON-like data<a class="headerlink" href="#mmjson-like-data" title="Permalink to this headline">¶</a></h3>
<p>Gemmi has a built-in support for mmJSON and comes with
a <a class="reference internal" href="utils.html#json"><span class="std std-ref">converter</span></a> utility, but just as an excercise
let us convert mmJSON to mmCIF in Python:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">gemmi</span> <span class="kn">import</span> <span class="n">cif</span>

<span class="n">file_in</span><span class="p">,</span> <span class="n">file_out</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_in</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">json_data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">json_data</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>  <span class="c1"># data_1ABC</span>
<span class="p">(</span><span class="n">block_name</span><span class="p">,</span> <span class="n">block_data</span><span class="p">),</span> <span class="o">=</span> <span class="n">json_data</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
<span class="k">assert</span> <span class="n">block_name</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;data_&#39;</span><span class="p">)</span>
<span class="c1"># Now block_data is a dictionary that maps category names to dictionaries</span>
<span class="c1"># that in turn map column names to lists with values.</span>
<span class="n">doc</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">Document</span><span class="p">()</span>
<span class="n">block</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_new_block</span><span class="p">(</span><span class="n">block_name</span><span class="p">[</span><span class="mi">5</span><span class="p">:])</span>
<span class="k">for</span> <span class="n">cat</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">block_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">block</span><span class="o">.</span><span class="n">set_mmcif_category</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">cat</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
<span class="n">doc</span><span class="o">.</span><span class="n">write_file</span><span class="p">(</span><span class="n">file_out</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="chemical-component-dictionary">
<span id="ccd-example"></span><h3>Chemical Component Dictionary<a class="headerlink" href="#chemical-component-dictionary" title="Permalink to this headline">¶</a></h3>
<p>For something a bit different, let us look at the data from
the <code class="file docutils literal notranslate"><span class="pre">components.cif</span></code> from <a class="reference external" href="https://www.wwpdb.org/data/ccd">CCD</a>.
This file describes all the monomers (residues,
ligands, solvent molecules) from the PDB entries.</p>
<p>As an exercise, let us check heavy atoms in selenomethionine:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">gemmi</span> <span class="kn">import</span> <span class="n">cif</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cif</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;components.cif&#39;</span><span class="p">)</span>  
<span class="go">&lt;gemmi.cif.Document with ... blocks (000, 001, 002...)&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span><span class="o">.</span><span class="n">find_block</span><span class="p">(</span><span class="s1">&#39;MSE&#39;</span><span class="p">)</span>
<span class="go">&lt;gemmi.cif.Block MSE&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;_chem_comp_atom.&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;atom_id&#39;</span><span class="p">,</span> <span class="s1">&#39;type_symbol&#39;</span><span class="p">])</span>
<span class="go">&lt;gemmi.cif.Table 20 x 2&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="s1">&#39;:&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">a</span><span class="p">)</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">_</span> <span class="k">if</span> <span class="n">a</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;H&#39;</span><span class="p">]</span>
<span class="go">[&#39;N:N&#39;, &#39;CA:C&#39;, &#39;C:C&#39;, &#39;O:O&#39;, &#39;OXT:O&#39;, &#39;CB:C&#39;, &#39;CG:C&#39;, &#39;SE:SE&#39;, &#39;CE:C&#39;]</span>
</pre></div>
</div>
<p>One may wonder what is the heaviest CCD component:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ccd</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;components.cif&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">max</span><span class="p">(</span><span class="n">ccd</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">b</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">b</span><span class="o">.</span><span class="n">find_value</span><span class="p">(</span><span class="s1">&#39;_chem_comp.formula_weight&#39;</span><span class="p">)))</span>
<span class="go">&lt;gemmi.cif.Block WO2&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span><span class="o">.</span><span class="n">find_value</span><span class="p">(</span><span class="s1">&#39;_chem_comp.formula&#39;</span><span class="p">)</span>
<span class="go">&#39;&quot;O62 P2 W18&quot;&#39;</span>
</pre></div>
</div>
<p>Or which one has the largest number of heavy atoms:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">el_tag</span> <span class="o">=</span> <span class="s1">&#39;_chem_comp_atom.type_symbol&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">max</span><span class="p">(</span><span class="n">ccd</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">b</span><span class="p">:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">el</span> <span class="o">!=</span> <span class="s1">&#39;H&#39;</span> <span class="k">for</span> <span class="n">el</span> <span class="ow">in</span> <span class="n">b</span><span class="o">.</span><span class="n">find_values</span><span class="p">(</span><span class="n">el_tag</span><span class="p">)))</span>
<span class="go">&lt;gemmi.cif.Block X12&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span><span class="o">.</span><span class="n">find_value</span><span class="p">(</span><span class="s1">&#39;_chem_comp.formula&#39;</span><span class="p">)</span>
<span class="go">&#39;&quot;C96 H153 N31 O25&quot;&#39;</span>
</pre></div>
</div>
<p>The <code class="file docutils literal notranslate"><span class="pre">components.cif</span></code> file is big, so we may want to split it into
multiple file. As an example, here we create a new document with only
one block, and we write it to a file:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">block</span> <span class="o">=</span> <span class="n">ccd</span><span class="p">[</span><span class="s1">&#39;X12&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">Document</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">add_copied_block</span><span class="p">(</span><span class="n">ccd</span><span class="p">[</span><span class="s1">&#39;X12&#39;</span><span class="p">])</span>
<span class="go">&lt;gemmi.cif.Block X12&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">write_file</span><span class="p">(</span><span class="s1">&#39;X12.cif&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>In the next example we delete things we do not need.
Let us write only components on letter A to a new file.
Additionally, we remove the descriptor category:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ccd</span> <span class="o">=</span> <span class="n">cif</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;components.cif&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">ccd</span><span class="p">)</span>  
<span class="go">25219</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">to_be_deleted</span> <span class="o">=</span> <span class="p">[</span><span class="n">n</span> <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">block</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ccd</span><span class="p">)</span> <span class="k">if</span> <span class="n">block</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;A&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">reversed</span><span class="p">(</span><span class="n">to_be_deleted</span><span class="p">):</span>
<span class="gp">... </span>  <span class="k">del</span> <span class="n">ccd</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">ccd</span><span class="p">)</span>  
<span class="go">991</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">block</span> <span class="ow">in</span> <span class="n">ccd</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">block</span><span class="o">.</span><span class="n">find_mmcif_category</span><span class="p">(</span><span class="s1">&#39;_pdbx_chem_comp_descriptor.&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">erase</span><span class="p">()</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ccd</span><span class="o">.</span><span class="n">write_file</span><span class="p">(</span><span class="s1">&#39;A.cif&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The examples here present low-level, generic handling of CCD as a CIF file.
If we would need to look into coordinates, it would be better to use
higher level representation of the chemical component, which is provided
by <a class="reference internal" href="mol.html#chemcomp"><span class="std std-ref">gemmi::ChemComp</span></a>.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="symmetry.html" class="btn btn-neutral float-right" title="Symmetry" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="install.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017-2020 Global Phasing Ltd

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>